<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>parkinsonsMetagenomicData</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">parkinsonsMetagenomicData</h1>



<div id="sample-metadata" class="section level1">
<h1>Sample Metadata</h1>
<p>Metadata for all samples present within
<code>parkinsonsMetagenomicData</code> is available through the included
<code>data.frame</code> <code>sampleMetadata</code>. Both curated and
uncurated features are included in this <code>data.frame</code>, with
uncurated features being prefixed by “uncurated_”. Curated features
include the following at this time:</p>
<ul>
<li>curation_id</li>
<li>study_name</li>
<li>sample_id</li>
<li>subject_id</li>
<li>target_condition</li>
<li>target_condition_ontology_term_id</li>
<li>control</li>
<li>control_ontology_term_id</li>
<li>age</li>
<li>age_group</li>
<li>age_group_ontology_term_id</li>
<li>age_unit</li>
<li>age_unit_ontology_term_id</li>
<li>sex</li>
<li>sex_ontology_term_id</li>
<li>disease</li>
<li>disease_ontology_term_id</li>
<li>curator</li>
<li>BioProject</li>
<li>BioSample</li>
<li>NCBI_accession</li>
<li>uuid</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>selected_samples <span class="ot">&lt;-</span> sampleMetadata <span class="sc">|&gt;</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>    <span class="fu">filter</span>(study_name <span class="sc">==</span> <span class="st">&quot;ZhangM_2023&quot;</span>) <span class="sc">|&gt;</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>    <span class="fu">select</span>(<span class="fu">where</span>(<span class="sc">~</span> <span class="sc">!</span><span class="fu">any</span>(<span class="fu">is.na</span>(.x))))</span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-10299551691e08827529" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-10299551691e08827529">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24"],["ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023"],["PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875"],["SAMN35813273","SAMN35813274","SAMN35813275","SAMN35813276","SAMN35813277","SAMN35813278","SAMN35813279","SAMN35813280","SAMN35813281","SAMN35813282","SAMN35813283","SAMN35813284","SAMN35813285","SAMN35813286","SAMN35813287","SAMN35813288","SAMN35813289","SAMN35813290","SAMN35813291","SAMN35813292","SAMN35813293","SAMN35813294","SAMN35813295","SAMN35813296"],["SRR25109817","SRR25109816","SRR25109805","SRR25109800","SRR25109799","SRR25109798","SRR25109797","SRR25109796","SRR25109795","SRR25109794","SRR25109815","SRR25109814","SRR25109813","SRR25109812","SRR25109811","SRR25109810","SRR25109809","SRR25109808","SRR25109807","SRR25109806","SRR25109804","SRR25109803","SRR25109802","SRR25109801"],["0807eb2a-a15e-4647-8e19-2600d8fda378","e0fbb54f-0249-4917-a4d7-bd68acb89c62","25172837-2849-4db3-be91-d54d6a815d00","39ddb5e7-97f6-4d3c-812b-9653b03f99b3","7b152a7d-e244-4e2b-b924-7195c7ecfb10","dd30f93b-7999-47a4-93fb-21971b899939","1406666f-04a8-43c9-983b-4ed62fd6da4a","fe3de3ca-3a14-4bd8-ae1c-0dad69edc9cd","8707e374-5ddb-4220-8cbf-364b8b0e7be1","22848a9c-66a6-4993-9058-cb6464edb42f","08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a","9baef0b2-93d2-4a40-8082-d357c7f8156a","09a9303d-d87d-4556-9672-04cbbcaf3d37","ac9f3532-90d8-412c-9c80-491037f0bcc2","eda61949-02dc-40ae-8dbe-bea2add85a52","1f007260-be6c-4a21-800a-ad9c36129a0d","e47a59bb-443a-405f-9c5d-02659d80e9e5","b3eaf3ab-43ef-4830-ab6d-12bafed3c61e","28f7352f-fe23-4003-93e1-41f4fedc6232","b07e2362-5851-4181-ba9a-15d9109ee4dd","677be4e3-722b-4e43-bd5a-36d8fbed6f86","0c817272-f873-475f-a401-dfe46a679a9f","7a3945d9-21bb-434a-9a4e-bfcdeb6194de","56aa2ad5-007d-407c-a644-48aac1e9a8f0"],["ZhangM_2023:F_PD1","ZhangM_2023:F_PD2","ZhangM_2023:F_PD3","ZhangM_2023:F_PD5","ZhangM_2023:F_PD7","ZhangM_2023:F_PD16","ZhangM_2023:F_PD19","ZhangM_2023:F_PD24","ZhangM_2023:F_PD27","ZhangM_2023:F_PD28","ZhangM_2023:F_PD29","ZhangM_2023:M_PD4","ZhangM_2023:M_PD6","ZhangM_2023:M_PD9","ZhangM_2023:M_PD10","ZhangM_2023:M_PD11","ZhangM_2023:M_PD13","ZhangM_2023:M_PD17","ZhangM_2023:M_PD18","ZhangM_2023:M_PD20","ZhangM_2023:M_PD21","ZhangM_2023:M_PD30","ZhangM_2023:M_PD31","ZhangM_2023:M_PD32"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease"],["NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845"],["Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case"],["NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152"],["Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease","Parkinson disease"],["NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845"],["Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long"],["SRR25109817","SRR25109816","SRR25109805","SRR25109800","SRR25109799","SRR25109798","SRR25109797","SRR25109796","SRR25109795","SRR25109794","SRR25109815","SRR25109814","SRR25109813","SRR25109812","SRR25109811","SRR25109810","SRR25109809","SRR25109808","SRR25109807","SRR25109806","SRR25109804","SRR25109803","SRR25109802","SRR25109801"],["WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS"],["281","274","282","277","280","287","291","289","290","292","291","280","280","290","282","286","286","290","289","290","290","293","292","291"],["5126812288","4637651878","4260703402","5073414290","4328969224","7837329037","5290309611","7625510674","8090442929","5777147435","5661210166","4332986236","5371153114","5654310289","4128355169","7327386738","7324599259","7868775948","7628779061","8040835028","5194000455","5795748371","5862491101","5693155117"],["PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875"],["SAMN35813273","SAMN35813274","SAMN35813275","SAMN35813276","SAMN35813277","SAMN35813278","SAMN35813279","SAMN35813280","SAMN35813281","SAMN35813282","SAMN35813283","SAMN35813284","SAMN35813285","SAMN35813286","SAMN35813287","SAMN35813288","SAMN35813289","SAMN35813290","SAMN35813291","SAMN35813292","SAMN35813293","SAMN35813294","SAMN35813295","SAMN35813296"],["1709080646","1540916207","1430374561","1688915874","1446219669","2588558077","1801807568","2520270145","2703434143","1922688037","1877184419","1448195135","1798553487","1915323502","1378503453","2419592743","2407455235","2593857558","2548175809","2672160392","1762093191","1933525398","1935544819","1895741776"],["THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY"],["public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public"],["fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra"],["gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3"],["gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1"],["SRX20862822","SRX20862823","SRX20862834","SRX20862839","SRX20862840","SRX20862841","SRX20862842","SRX20862843","SRX20862844","SRX20862845","SRX20862824","SRX20862825","SRX20862826","SRX20862827","SRX20862828","SRX20862829","SRX20862830","SRX20862831","SRX20862832","SRX20862833","SRX20862835","SRX20862836","SRX20862837","SRX20862838"],["China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China"],["Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia"],["China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou"],["Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE"],["RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM"],["METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC"],["human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome"],["ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA"],["2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z"],["2023-07-02T04:11:00Z","2023-07-02T04:11:00Z","2023-07-02T04:11:00Z","2023-07-02T04:10:00Z","2023-07-02T04:13:00Z","2023-07-02T04:18:00Z","2023-07-02T04:10:00Z","2023-07-02T04:12:00Z","2023-07-02T04:14:00Z","2023-07-02T04:13:00Z","2023-07-02T04:12:00Z","2023-07-02T04:11:00Z","2023-07-02T04:11:00Z","2023-07-02T04:14:00Z","2023-07-02T04:10:00Z","2023-07-02T04:12:00Z","2023-07-02T04:14:00Z","2023-07-02T04:18:00Z","2023-07-02T04:20:00Z","2023-07-02T04:16:00Z","2023-07-02T04:10:00Z","2023-07-02T04:11:00Z","2023-07-02T04:12:00Z","2023-07-02T04:12:00Z"],["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959"],["2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12"],["Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental"],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>study_name<\/th>\n      <th>BioProject<\/th>\n      <th>BioSample<\/th>\n      <th>NCBI_accession<\/th>\n      <th>uuid<\/th>\n      <th>curation_id<\/th>\n      <th>sample_id<\/th>\n      <th>subject_id<\/th>\n      <th>target_condition<\/th>\n      <th>target_condition_ontology_term_id<\/th>\n      <th>control<\/th>\n      <th>control_ontology_term_id<\/th>\n      <th>disease<\/th>\n      <th>disease_ontology_term_id<\/th>\n      <th>curator<\/th>\n      <th>uncurated_Run<\/th>\n      <th>uncurated_Assay.Type<\/th>\n      <th>uncurated_AvgSpotLen<\/th>\n      <th>uncurated_Bases<\/th>\n      <th>uncurated_BioProject<\/th>\n      <th>uncurated_BioSample<\/th>\n      <th>uncurated_Bytes<\/th>\n      <th>uncurated_Center.Name<\/th>\n      <th>uncurated_Consent<\/th>\n      <th>uncurated_DATASTORE.filetype<\/th>\n      <th>uncurated_DATASTORE.provider<\/th>\n      <th>uncurated_DATASTORE.region<\/th>\n      <th>uncurated_Experiment<\/th>\n      <th>uncurated_geo_loc_name_country<\/th>\n      <th>uncurated_geo_loc_name_country_continent<\/th>\n      <th>uncurated_geo_loc_name<\/th>\n      <th>uncurated_Instrument<\/th>\n      <th>uncurated_Library.Name<\/th>\n      <th>uncurated_LibraryLayout<\/th>\n      <th>uncurated_LibrarySelection<\/th>\n      <th>uncurated_LibrarySource<\/th>\n      <th>uncurated_Organism<\/th>\n      <th>uncurated_Platform<\/th>\n      <th>uncurated_ReleaseDate<\/th>\n      <th>uncurated_create_date<\/th>\n      <th>uncurated_version<\/th>\n      <th>uncurated_Sample.Name<\/th>\n      <th>uncurated_SRA.Study<\/th>\n      <th>uncurated_Collection_Date<\/th>\n      <th>uncurated_BioSampleModel<\/th>\n      <th>uncurated_HOST<\/th>\n      <th>uncurated_lat_lon<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"study_name","targets":1},{"name":"BioProject","targets":2},{"name":"BioSample","targets":3},{"name":"NCBI_accession","targets":4},{"name":"uuid","targets":5},{"name":"curation_id","targets":6},{"name":"sample_id","targets":7},{"name":"subject_id","targets":8},{"name":"target_condition","targets":9},{"name":"target_condition_ontology_term_id","targets":10},{"name":"control","targets":11},{"name":"control_ontology_term_id","targets":12},{"name":"disease","targets":13},{"name":"disease_ontology_term_id","targets":14},{"name":"curator","targets":15},{"name":"uncurated_Run","targets":16},{"name":"uncurated_Assay.Type","targets":17},{"name":"uncurated_AvgSpotLen","targets":18},{"name":"uncurated_Bases","targets":19},{"name":"uncurated_BioProject","targets":20},{"name":"uncurated_BioSample","targets":21},{"name":"uncurated_Bytes","targets":22},{"name":"uncurated_Center.Name","targets":23},{"name":"uncurated_Consent","targets":24},{"name":"uncurated_DATASTORE.filetype","targets":25},{"name":"uncurated_DATASTORE.provider","targets":26},{"name":"uncurated_DATASTORE.region","targets":27},{"name":"uncurated_Experiment","targets":28},{"name":"uncurated_geo_loc_name_country","targets":29},{"name":"uncurated_geo_loc_name_country_continent","targets":30},{"name":"uncurated_geo_loc_name","targets":31},{"name":"uncurated_Instrument","targets":32},{"name":"uncurated_Library.Name","targets":33},{"name":"uncurated_LibraryLayout","targets":34},{"name":"uncurated_LibrarySelection","targets":35},{"name":"uncurated_LibrarySource","targets":36},{"name":"uncurated_Organism","targets":37},{"name":"uncurated_Platform","targets":38},{"name":"uncurated_ReleaseDate","targets":39},{"name":"uncurated_create_date","targets":40},{"name":"uncurated_version","targets":41},{"name":"uncurated_Sample.Name","targets":42},{"name":"uncurated_SRA.Study","targets":43},{"name":"uncurated_Collection_Date","targets":44},{"name":"uncurated_BioSampleModel","targets":45},{"name":"uncurated_HOST","targets":46},{"name":"uncurated_lat_lon","targets":47}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="output-files" class="section level1">
<h1>Output files</h1>
<p>The UUID(s) of any sample(s) of interest may be then used to access
output files. Available data types and information about their full file
paths can be found with <code>output_file_types()</code></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>ftypes <span class="ot">&lt;-</span> <span class="fu">output_file_types</span>()</span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ef1c50dfb54156e310f6" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ef1c50dfb54156e310f6">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27"],["viral_clusters","relative_abundance","marker_abundance","marker_presence","genefamilies","genefamilies_cpm","genefamilies_relab","genefamilies_stratified","genefamilies_unstratified","genefamilies_cpm_stratified","genefamilies_relab_stratified","genefamilies_cpm_unstratified","genefamilies_relab_unstratified","pathabundance","pathabundance_cpm","pathabundance_relab","pathabundance_stratified","pathabundance_unstratified","pathabundance_cpm_stratified","pathabundance_relab_stratified","pathabundance_cpm_unstratified","pathabundance_relab_unstratified","pathcoverage_unstratified","pathcoverage_stratified","pathcoverage","fastqc","kneaddata_log"],["metaphlan_viruses_list.tsv.gz","metaphlan_unknown_list.tsv.gz","marker_abundance.tsv.gz","marker_presence.tsv.gz","out_genefamilies.tsv.gz","out_genefamilies_cpm.tsv.gz","out_genefamilies_relab.tsv.gz","out_genefamilies_stratified.tsv.gz","out_genefamilies_unstratified.tsv.gz","out_genefamilies_cpm_stratified.tsv.gz","out_genefamilies_relab_stratified.tsv.gz","out_genefamilies_cpm_unstratified.tsv.gz","out_genefamilies_relab_unstratified.tsv.gz","out_pathabundance.tsv.gz","out_pathabundance_cpm.tsv.gz","out_pathabundance_relab.tsv.gz","out_pathabundance_stratified.tsv.gz","out_pathabundance_unstratified.tsv.gz","out_pathabundance_cpm_stratified.tsv.gz","out_pathabundance_relab_stratified.tsv.gz","out_pathabundance_cpm_unstratified.tsv.gz","out_pathabundance_relab_unstratified.tsv.gz","out_pathcoverage_unstratified.tsv.gz","out_pathcoverage_stratified.tsv.gz","out_pathcoverage.tsv.gz","fastqc_data.txt","out_kneaddata.log"],["metaphlan_lists/","metaphlan_lists/","metaphlan_markers/","metaphlan_markers/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","humann/","fasterq_dump/out_sample_fastqc/","kneaddata/kneaddata_output/"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>data_type<\/th>\n      <th>file_name<\/th>\n      <th>subdir<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"data_type","targets":1},{"name":"file_name","targets":2},{"name":"subdir","targets":3}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script>
<p>At the moment, only the <code>metaphlan_lists</code> data types,
<code>viral_clusters</code> and <code>relative_abundance</code>, have
parsing functions for automatically loading into SummarizedExperiment
objects.</p>
</div>
<div id="google-bucket-setup" class="section level1">
<h1>Google Bucket Setup</h1>
<p>Output files are contained in the private Google Cloud Bucket
<code>gs://metagenomics-mac</code> at the moment, the user must
authenticate with a service account keyfile in the following way:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>googleCloudStorageR<span class="sc">::</span><span class="fu">gcs_auth</span>(<span class="st">&quot;full/path/to/keyfile.json&quot;</span>)</span></code></pre></div>
<p>To obtain this keyfile, the owner of a service account affiliated
with the Google Cloud Project containing
<code>gs://metagenomics-mac</code> must create it according to the
process detailed in the Google Cloud Guide “<a href="https://cloud.google.com/iam/docs/keys-create-delete">Create and
delete service account keys</a>”.</p>
<p>Additionally, the default bucket for <code>GoogleCloudStorageR</code>
must be set to the Google Bucket name, in this case
<code>metagenomics-mac</code>. This is done automatically upon loading
the package, but if an error occurs can be manually done through the
following:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>googleCloudStorageR<span class="sc">::</span><span class="fu">gcs_global_bucket</span>(<span class="st">&quot;metagenomics-mac&quot;</span>)</span></code></pre></div>
</div>
<div id="data-retrieval" class="section level1">
<h1>Data Retrieval</h1>
<p>All objects stored in <code>gs://metagenomics-mac</code> can then be
viewed with the following. This operation can take several minutes due
to the number of objects stored in
<code>gs://metagenomics-mac</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>googleCloudStorageR<span class="sc">::</span><span class="fu">gcs_list_objects</span>()</span></code></pre></div>
<p>Alternatively, <code>listMetagenomicData</code> will provide a table
of objects that are compatible with the
<code>cacheMetagenomicData</code> streamlined downloading and caching
function.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>file_tbl <span class="ot">&lt;-</span> <span class="fu">listMetagenomicData</span>()</span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-187b419a909c21128781" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-187b419a909c21128781">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10"],["3bf45362-1b3b-4d13-90f9-747d4f3c0667","78b2bef7-173b-4191-a9f1-f6176e0812ff","ae271b6f-9f1b-4a64-8764-758e6f6c627a","35c273e7-a7e4-4e24-b307-eff8115e4dfb","1da57f64-20c6-4db0-a2d8-fdaa9fc10c26","40fcafa2-7823-4086-902c-76edd44c8ad7","6e095f32-fd3d-4990-bdb6-df5d5f44941f","5582ab8f-10e0-41d6-b851-b12c8a67955b","d4217ace-a87d-4d21-95d8-58c434f6f279","0fbd0449-e9b8-48be-9511-9bed7525ec19"],["marker_abundance","genefamilies_relab_stratified","pathcoverage","genefamilies","viral_clusters","marker_abundance","pathabundance","viral_clusters","pathabundance_relab_unstratified","genefamilies_relab"],["results/cMDv4/3bf45362-1b3b-4d13-90f9-747d4f3c0667/metaphlan_markers/marker_abundance.tsv.gz","results/cMDv4/78b2bef7-173b-4191-a9f1-f6176e0812ff/humann/out_genefamilies_relab_stratified.tsv.gz","results/cMDv4/ae271b6f-9f1b-4a64-8764-758e6f6c627a/humann/out_pathcoverage.tsv.gz","results/cMDv4/35c273e7-a7e4-4e24-b307-eff8115e4dfb/humann/out_genefamilies.tsv.gz","results/cMDv4/1da57f64-20c6-4db0-a2d8-fdaa9fc10c26/metaphlan_lists/metaphlan_viruses_list.tsv.gz","results/cMDv4/40fcafa2-7823-4086-902c-76edd44c8ad7/metaphlan_markers/marker_abundance.tsv.gz","results/cMDv4/6e095f32-fd3d-4990-bdb6-df5d5f44941f/humann/out_pathabundance.tsv.gz","results/cMDv4/5582ab8f-10e0-41d6-b851-b12c8a67955b/metaphlan_lists/metaphlan_viruses_list.tsv.gz","results/cMDv4/d4217ace-a87d-4d21-95d8-58c434f6f279/humann/out_pathabundance_relab_unstratified.tsv.gz","results/cMDv4/0fbd0449-e9b8-48be-9511-9bed7525ec19/humann/out_genefamilies_relab.tsv.gz"],["406.4 Kb","2.8 Mb","38.8 Kb","4.7 Mb","1.2 Kb","537.2 Kb","15.3 Kb","452 bytes","8.1 Kb","4.4 Mb"],["2025-05-06T00:19:01Z","2025-05-26T13:53:55Z","2025-05-06T00:31:53Z","2025-05-06T00:18:20Z","2025-05-06T00:15:37Z","2025-05-06T00:19:37Z","2025-05-27T03:44:34Z","2025-05-25T09:13:44Z","2025-05-06T00:36:24Z","2025-05-27T11:13:45Z"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>UUID<\/th>\n      <th>data_type<\/th>\n      <th>gcb_object<\/th>\n      <th>size<\/th>\n      <th>updated<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"UUID","targets":1},{"name":"data_type","targets":2},{"name":"gcb_object","targets":3},{"name":"size","targets":4},{"name":"updated","targets":5}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script>
<p>To access output files, use <code>cacheMetagenomicData</code>. This
function takes UUID and data type arguments, downloads the corresponding
output files, and stores them in a local cache. If the same files are
requested again through this function, they will not be re-downloaded
unless explicitly specified, in order to reduce excessive downloads.
<code>cacheMetagenomicData</code> returns a tibble of cached file paths
and cache IDs for each requested file.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>cache_tbl <span class="ot">&lt;-</span> <span class="fu">cacheMetagenomicData</span>(<span class="at">uuids =</span> selected_samples<span class="sc">$</span>uuid,</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>                                  <span class="at">data_type =</span> <span class="st">&quot;relative_abundance&quot;</span>)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/0807eb2a-a15e-4647-8e19-2600d8fda378/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/e0fbb54f-0249-4917-a4d7-bd68acb89c62/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/25172837-2849-4db3-be91-d54d6a815d00/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/39ddb5e7-97f6-4d3c-812b-9653b03f99b3/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/7b152a7d-e244-4e2b-b924-7195c7ecfb10/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/dd30f93b-7999-47a4-93fb-21971b899939/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/1406666f-04a8-43c9-983b-4ed62fd6da4a/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/fe3de3ca-3a14-4bd8-ae1c-0dad69edc9cd/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/8707e374-5ddb-4220-8cbf-364b8b0e7be1/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/22848a9c-66a6-4993-9058-cb6464edb42f/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/9baef0b2-93d2-4a40-8082-d357c7f8156a/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/09a9303d-d87d-4556-9672-04cbbcaf3d37/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/ac9f3532-90d8-412c-9c80-491037f0bcc2/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/eda61949-02dc-40ae-8dbe-bea2add85a52/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/1f007260-be6c-4a21-800a-ad9c36129a0d/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/e47a59bb-443a-405f-9c5d-02659d80e9e5/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/b3eaf3ab-43ef-4830-ab6d-12bafed3c61e/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/28f7352f-fe23-4003-93e1-41f4fedc6232/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/b07e2362-5851-4181-ba9a-15d9109ee4dd/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/677be4e3-722b-4e43-bd5a-36d8fbed6f86/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/0c817272-f873-475f-a401-dfe46a679a9f/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/7a3945d9-21bb-434a-9a4e-bfcdeb6194de/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a><span class="co">#| Resource with rname = &#39;results/cMDv4/56aa2ad5-007d-407c-a644-48aac1e9a8f0/metaphlan_lists/metaphlan_unknown_list.tsv.gz&#39; found in cache, proceeding with most recent version.</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-6e503f92d7ebf841cf93" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-6e503f92d7ebf841cf93">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24"],["0807eb2a-a15e-4647-8e19-2600d8fda378","e0fbb54f-0249-4917-a4d7-bd68acb89c62","25172837-2849-4db3-be91-d54d6a815d00","39ddb5e7-97f6-4d3c-812b-9653b03f99b3","7b152a7d-e244-4e2b-b924-7195c7ecfb10","dd30f93b-7999-47a4-93fb-21971b899939","1406666f-04a8-43c9-983b-4ed62fd6da4a","fe3de3ca-3a14-4bd8-ae1c-0dad69edc9cd","8707e374-5ddb-4220-8cbf-364b8b0e7be1","22848a9c-66a6-4993-9058-cb6464edb42f","08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a","9baef0b2-93d2-4a40-8082-d357c7f8156a","09a9303d-d87d-4556-9672-04cbbcaf3d37","ac9f3532-90d8-412c-9c80-491037f0bcc2","eda61949-02dc-40ae-8dbe-bea2add85a52","1f007260-be6c-4a21-800a-ad9c36129a0d","e47a59bb-443a-405f-9c5d-02659d80e9e5","b3eaf3ab-43ef-4830-ab6d-12bafed3c61e","28f7352f-fe23-4003-93e1-41f4fedc6232","b07e2362-5851-4181-ba9a-15d9109ee4dd","677be4e3-722b-4e43-bd5a-36d8fbed6f86","0c817272-f873-475f-a401-dfe46a679a9f","7a3945d9-21bb-434a-9a4e-bfcdeb6194de","56aa2ad5-007d-407c-a644-48aac1e9a8f0"],["relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance"],["results/cMDv4/0807eb2a-a15e-4647-8e19-2600d8fda378/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/e0fbb54f-0249-4917-a4d7-bd68acb89c62/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/25172837-2849-4db3-be91-d54d6a815d00/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/39ddb5e7-97f6-4d3c-812b-9653b03f99b3/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/7b152a7d-e244-4e2b-b924-7195c7ecfb10/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/dd30f93b-7999-47a4-93fb-21971b899939/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/1406666f-04a8-43c9-983b-4ed62fd6da4a/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/fe3de3ca-3a14-4bd8-ae1c-0dad69edc9cd/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/8707e374-5ddb-4220-8cbf-364b8b0e7be1/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/22848a9c-66a6-4993-9058-cb6464edb42f/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/9baef0b2-93d2-4a40-8082-d357c7f8156a/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/09a9303d-d87d-4556-9672-04cbbcaf3d37/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/ac9f3532-90d8-412c-9c80-491037f0bcc2/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/eda61949-02dc-40ae-8dbe-bea2add85a52/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/1f007260-be6c-4a21-800a-ad9c36129a0d/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/e47a59bb-443a-405f-9c5d-02659d80e9e5/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/b3eaf3ab-43ef-4830-ab6d-12bafed3c61e/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/28f7352f-fe23-4003-93e1-41f4fedc6232/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/b07e2362-5851-4181-ba9a-15d9109ee4dd/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/677be4e3-722b-4e43-bd5a-36d8fbed6f86/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/0c817272-f873-475f-a401-dfe46a679a9f/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/7a3945d9-21bb-434a-9a4e-bfcdeb6194de/metaphlan_lists/metaphlan_unknown_list.tsv.gz","results/cMDv4/56aa2ad5-007d-407c-a644-48aac1e9a8f0/metaphlan_lists/metaphlan_unknown_list.tsv.gz"],["BFC6786","BFC6787","BFC6788","BFC6789","BFC6790","BFC6791","BFC6792","BFC6793","BFC6794","BFC6795","BFC6796","BFC6797","BFC6798","BFC6799","BFC6800","BFC6801","BFC6802","BFC6803","BFC6804","BFC6805","BFC6806","BFC6807","BFC6808","BFC6809"],["/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed766d3d8b.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed316de998.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed2f25ff85.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32edb2aefc9.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed750e9261.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed758c30c8.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed40ac0d62.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed6a5fa104.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32edc5bde52.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed68392cbc.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed4a29651f.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed123174ed.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed2be3f7a0.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed62c58fc1.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed78c3747a.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed277d2f3b.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed5e5156eb.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed42514a5e.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed48310172.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed5f1f853b.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32edded4142.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed168269a.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed5feec164.tsv.gz","/home/kaelyn/.cache/R/parkinsonsMetagenomicData/a32ed1b6bca0f.tsv.gz"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>UUID<\/th>\n      <th>data_type<\/th>\n      <th>gcb_object<\/th>\n      <th>cache_id<\/th>\n      <th>cache_path<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"UUID","targets":1},{"name":"data_type","targets":2},{"name":"gcb_object","targets":3},{"name":"cache_id","targets":4},{"name":"cache_path","targets":5}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="data-handling" class="section level1">
<h1>Data Handling</h1>
<div id="automatic-experiment-setup-metaphlan-output-files-only" class="section level2">
<h2>Automatic Experiment Setup (MetaPhlAn output files only)</h2>
<p>The above table can then be supplied to
<code>loadMetagenomicData</code> and the cached files will be parsed
into a single SummarizedExperiment object with sample metadata. At this
point, only the <code>metaphlan_lists</code> data types,
<code>viral_clusters</code> and <code>relative_abundance</code>, are
compatible with this function.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>merged_experiment <span class="ot">&lt;-</span> <span class="fu">loadMetagenomicData</span>(cache_tbl)</span></code></pre></div>
<pre><code>#| class: SummarizedExperiment 
#| dim: 2915 24 
#| metadata(0):
#| assays(1): relative_abundance
#| rownames(2915): UNCLASSIFIED k__Bacteria ...
#|   k__Bacteria|p__Firmicutes|c__CFGB3009|o__OFGB3009|f__FGB3009|g__GGB31234|s__GGB31234_SGB14869|t__SGB14869
#|   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Eubacteriales_unclassified|g__Eubacteriales_unclassified|s__Clostridiales_bacterium_CHKCI006|t__SGB7261
#| rowData names(2): ncbi_tax_id additional_species
#| colnames(24): 0807eb2a-a15e-4647-8e19-2600d8fda378
#|   e0fbb54f-0249-4917-a4d7-bd68acb89c62 ...
#|   7a3945d9-21bb-434a-9a4e-bfcdeb6194de
#|   56aa2ad5-007d-407c-a644-48aac1e9a8f0
#| colData names(521): study_name BioProject ... date_of_birth cage</code></pre>
</div>
<div id="stepwise-experiment-setup" class="section level2">
<h2>Stepwise Experiment Setup</h2>
<p>Alternatively, we can parse the files, add metadata, and merge the
SummarizedExperiment objects in separate steps.</p>
<p><code>parse_metaphlan_list</code> completes the first step for files
with ‘data_type’ equal to “relative_abundance” or “viral_clusters”.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>parsed_rel_ab_list <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">&quot;list&quot;</span>, <span class="fu">nrow</span>(cache_tbl))</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">names</span>(parsed_rel_ab_list) <span class="ot">&lt;-</span> cache_tbl<span class="sc">$</span>UUID</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(cache_tbl)) {</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>    parsed_rel_ab_list[[i]] <span class="ot">&lt;-</span> <span class="fu">parse_metaphlan_list</span>(<span class="at">sample_id =</span> cache_tbl<span class="sc">$</span>UUID[i],</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>                                                <span class="at">file_path =</span> cache_tbl<span class="sc">$</span>cache_path[i],</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>                                                <span class="at">data_type =</span> cache_tbl<span class="sc">$</span>data_type[i])</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>    </span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>}</span></code></pre></div>
<p>Once the files have been loaded as SummarizedExperiment objects,
matching assays from multiple samples can be merged together with
<code>mergeExperiments</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>merged_rel_abs <span class="ot">&lt;-</span> <span class="fu">mergeExperiments</span>(parsed_rel_ab_list)</span></code></pre></div>
<pre><code>#| class: SummarizedExperiment 
#| dim: 2915 24 
#| metadata(0):
#| assays(1): relative_abundance
#| rownames(2915): UNCLASSIFIED k__Bacteria ...
#|   k__Bacteria|p__Firmicutes|c__CFGB3009|o__OFGB3009|f__FGB3009|g__GGB31234|s__GGB31234_SGB14869|t__SGB14869
#|   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Eubacteriales_unclassified|g__Eubacteriales_unclassified|s__Clostridiales_bacterium_CHKCI006|t__SGB7261
#| rowData names(2): ncbi_tax_id additional_species
#| colnames(24): 0807eb2a-a15e-4647-8e19-2600d8fda378
#|   e0fbb54f-0249-4917-a4d7-bd68acb89c62 ...
#|   7a3945d9-21bb-434a-9a4e-bfcdeb6194de
#|   56aa2ad5-007d-407c-a644-48aac1e9a8f0
#| colData names(0):</code></pre>
<p>The corresponding metadata from sampleMetadata is then added as
colData with the function <code>add_metadata</code>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>rel_abs_with_metadata <span class="ot">&lt;-</span> <span class="fu">add_metadata</span>(<span class="at">sample_ids =</span> <span class="fu">colnames</span>(merged_rel_abs),</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>                                      <span class="at">id_col =</span> <span class="st">&quot;uuid&quot;</span>,</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>                                      <span class="at">experiment =</span> merged_rel_abs)</span></code></pre></div>
<pre><code>#| class: SummarizedExperiment 
#| dim: 2915 24 
#| metadata(0):
#| assays(1): relative_abundance
#| rownames(2915): UNCLASSIFIED k__Bacteria ...
#|   k__Bacteria|p__Firmicutes|c__CFGB3009|o__OFGB3009|f__FGB3009|g__GGB31234|s__GGB31234_SGB14869|t__SGB14869
#|   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Eubacteriales_unclassified|g__Eubacteriales_unclassified|s__Clostridiales_bacterium_CHKCI006|t__SGB7261
#| rowData names(2): ncbi_tax_id additional_species
#| colnames(24): 0807eb2a-a15e-4647-8e19-2600d8fda378
#|   e0fbb54f-0249-4917-a4d7-bd68acb89c62 ...
#|   7a3945d9-21bb-434a-9a4e-bfcdeb6194de
#|   56aa2ad5-007d-407c-a644-48aac1e9a8f0
#| colData names(521): study_name BioProject ... date_of_birth cage</code></pre>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
