<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Working With Data in Parquet Format • parkinsonsMetagenomicData</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Working With Data in Parquet Format">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">parkinsonsMetagenomicData</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/parquet.html">Working With Data in Parquet Format</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="parquet_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="parquet_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="parquet_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="parquet_files/datatables-binding-0.34.0/datatables.js"></script><link href="parquet_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="parquet_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="parquet_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="parquet_files/dt-ext-responsive-1.13.6/css/responsive.dataTables.min.css" rel="stylesheet">
<script src="parquet_files/dt-ext-responsive-1.13.6/js/dataTables.responsive.min.js"></script><link href="parquet_files/crosstalk-1.2.2/css/crosstalk.min.css" rel="stylesheet">
<script src="parquet_files/crosstalk-1.2.2/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Working With Data in Parquet Format</h1>
            
      

      <div class="d-none name"><code>parquet.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2">
<h2 id="parquet-file-overview-and-setup">Parquet File Overview and Setup<a class="anchor" aria-label="anchor" href="#parquet-file-overview-and-setup"></a>
</h2>
<div class="section level3">
<h3 id="hugging-face">Hugging Face<a class="anchor" aria-label="anchor" href="#hugging-face"></a>
</h3>
<p>Individual pipeline output files have been combined into parquet
files and hosted in the <a href="https://huggingface.co/datasets/waldronlab/metagenomics_mac" class="external-link">metagenomics_mac</a>
repo on Hugging Face. Smaller versions of those same files, comprising
only 10 samples per file, are available in the Hugging Face repo <a href="https://huggingface.co/datasets/waldronlab/metagenomics_mac_examples" class="external-link">metagenomics_mac_examples</a>.
These are publicly accessible, and are able to be easily read using
DuckDB.</p>
</div>
<div class="section level3">
<h3 id="parquet-creation">Parquet Creation<a class="anchor" aria-label="anchor" href="#parquet-creation"></a>
</h3>
<p>To create the individual parquet files, one needs credentials to
access the <code>gs://metagenomics-mac</code> Google Bucket.
Instructions on obtaining these can be found in the <a href="https://asap-mac.github.io/parkinsonsMetagenomicData/docs/vignettes/google_cloud_storage.html">Google
Cloud Storage vignette</a>. These credentials are then substituted into
the scripts found in the repo <a href="https://github.com/ASAP-MAC/parquet_generation" class="external-link">parquet_generation</a>.
These scripts perform a number of transformations that increase
searchability before storing the combined data for each data type in
parquet files.</p>
</div>
<div class="section level3">
<h3 id="duckdb-in-r">DuckDB in R<a class="anchor" aria-label="anchor" href="#duckdb-in-r"></a>
</h3>
<p>DuckDB is very easy to use in R through the <code>duckdb</code> R
package. The <code>DBI</code> and <code>dplyr</code>/<code>dbplyr</code>
packages combine with it to provide a streamlined way to work with
remote data by selectively querying it before bringing it into your R
session.</p>
<p>Relevant tools:</p>
<ul>
<li><a href="https://duckdb.org/docs/stable/clients/r.html" class="external-link">DuckDB R
Client</a></li>
<li><a href="https://dbi.r-dbi.org/" class="external-link">DBI</a></li>
<li>
<a href="https://dplyr.tidyverse.org/" class="external-link">dplyr</a>/<a href="https://dbplyr.tidyverse.org/" class="external-link">dbplyr</a>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="standard-workflow">Standard Workflow<a class="anchor" aria-label="anchor" href="#standard-workflow"></a>
</h2>
<p>In the standard workflow, we use the main wrapper function
<code>returnSamples</code>. This function takes tables with sample and
feature information as well as a remote repo name or a vector of paths
to locally stored parquet files and retrieves the relevant data as a
TreeSummarizedExperiment.</p>
<p>Here is a brief overview of each argument to be provided to
<code>returnSamples</code>, please call <code><a href="../reference/returnSamples.html">?returnSamples</a></code> for
additional info.</p>
<ul>
<li>
<strong>data_type</strong>: the output file type of interest</li>
<li>
<strong>sample_data</strong>: a table of sample metadata, used to
specify which samples to retrieve data for</li>
<li>
<strong>feature_data</strong>: a table of feature data, used to
specify how to filter the raw data</li>
<li>
<strong>repo</strong>: the identifier of a remote repo where the raw
parquet files are stored</li>
<li>
<strong>local_files</strong>: paths of locally stored parquet files,
as an alternative to retrieval from a remote repo</li>
<li>
<strong>include_empty_samples</strong>: sometimes none of the
features specified in <code>feature_data</code> are found in one or more
samples specified in <code>sample_data</code>. Should the samples still
be included in the result with NA values for each feature (TRUE) or
should they be omitted (FALSE)? Omitting these samples may increase
response speed when accessing remote repos.</li>
<li>
<strong>dry_run</strong>: a dry run returns a
<code>tbl_duckdb_connection</code> object that contains all of the SQL
code necessary to return the requested data. This SQL can be viewed by
passing the object into <code><a href="https://dplyr.tidyverse.org/reference/explain.html" class="external-link">dplyr::show_query()</a></code>. This is useful
for evaluating query efficiency prior to actually executing.</li>
</ul>
<div class="section level3">
<h3 id="file-selection">File Selection<a class="anchor" aria-label="anchor" href="#file-selection"></a>
</h3>
<p>Both remote and local parquet files can be queried, though not at the
same time. <code><a href="../reference/get_repo_info.html">get_repo_info()</a></code> gives the names and URLs of the
available repos, and <code><a href="../reference/get_hf_parquet_urls.html">get_hf_parquet_urls()</a></code> gives
information on the files contained in those repos. Those files can also
be downloaded separately and then provided to the
<code>local_files</code> argument of <code><a href="../reference/returnSamples.html">returnSamples()</a></code>. This
may be desirable if internet connection is unstable or limited, or if a
particular query runs into the repo’s rate limits.</p>
<p>At this point, we also want to make a note of which
<code>data_type</code> values we are interested in. The
<code>data_type</code> associated with each parquet file is listed in
the output of <code><a href="../reference/get_hf_parquet_urls.html">get_hf_parquet_urls()</a></code>. For this example, we
will be looking at the ‘relative_abundance’ data type, which corresponds
to the files “relative_abundance_uuid.parquet” and
“relative_abundance_clade_name_species.parquet”. The difference between
these two files is the column by which they are sorted:
“relative_abundance_uuid.parquet” is sorted by the ‘uuid’ column, while
“relative_abundance_clade_name_species.parquet” is sorted by the
‘clade_name_species’ column. Internal functions will choose which one to
use, we will simply provide “relative_abundance” as our chosen data
type. If you are downloading the files locally to avoid rate limits, you
would download all of the files associated with your chosen data
type.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_repo_info.html">get_repo_info</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2"],["waldronlab/metagenomics_mac","waldronlab/metagenomics_mac_examples"],["https://huggingface.co/datasets/waldronlab/metagenomics_mac/tree/main","https://huggingface.co/datasets/waldronlab/metagenomics_mac_examples/tree/main"],["Y","N"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>repo_name<\/th>\n      <th>repo_url<\/th>\n      <th>default<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"repo_name","targets":1},{"name":"repo_url","targets":2},{"name":"default","targets":3}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_hf_parquet_urls.html">get_hf_parquet_urls</a></span><span class="op">(</span>repo_name <span class="op">=</span> <span class="st">"waldronlab/metagenomics_mac"</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55"],["clade_name_ref.parquet","gene_family_ref.parquet","genefamilies_cpm_gene_family_uniref.parquet","genefamilies_cpm_stratified_gene_family_uniref.parquet","genefamilies_cpm_stratified_uuid.parquet","genefamilies_cpm_unstratified_gene_family.parquet","genefamilies_cpm_unstratified_uuid.parquet","genefamilies_cpm_uuid.parquet","genefamilies_gene_family_uniref.parquet","genefamilies_relab_gene_family_uniref.parquet","genefamilies_relab_stratified_gene_family_uniref.parquet","genefamilies_relab_stratified_uuid.parquet","genefamilies_relab_unstratified_gene_family.parquet","genefamilies_relab_unstratified_uuid.parquet","genefamilies_relab_uuid.parquet","genefamilies_stratified_gene_family_uniref.parquet","genefamilies_stratified_uuid.parquet","genefamilies_unstratified_gene_family.parquet","genefamilies_unstratified_uuid.parquet","genefamilies_uuid.parquet","genome_name_ref.parquet","marker_abundance_uniref.parquet","marker_abundance_uuid.parquet","marker_presence_uniref.parquet","marker_presence_uuid.parquet","pathabundance_cpm_pathway_uniref.parquet","pathabundance_cpm_stratified_pathway_uniref.parquet","pathabundance_cpm_stratified_uuid.parquet","pathabundance_cpm_unstratified_pathway.parquet","pathabundance_cpm_unstratified_uuid.parquet","pathabundance_cpm_uuid.parquet","pathabundance_pathway_uniref.parquet","pathabundance_relab_pathway_uniref.parquet","pathabundance_relab_stratified_pathway_uniref.parquet","pathabundance_relab_stratified_uuid.parquet","pathabundance_relab_unstratified_pathway.parquet","pathabundance_relab_unstratified_uuid.parquet","pathabundance_relab_uuid.parquet","pathabundance_stratified_pathway_uniref.parquet","pathabundance_stratified_uuid.parquet","pathabundance_unstratified_pathway.parquet","pathabundance_unstratified_uuid.parquet","pathabundance_uuid.parquet","pathcoverage_pathway_uniref.parquet","pathcoverage_stratified_pathway_uniref.parquet","pathcoverage_stratified_uuid.parquet","pathcoverage_unstratified_pathway.parquet","pathcoverage_unstratified_uuid.parquet","pathcoverage_uuid.parquet","pathway_ref.parquet","relative_abundance_clade_name_species.parquet","relative_abundance_uuid.parquet","uniref_marker_ref.parquet","viral_clusters_genome_name.parquet","viral_clusters_uuid.parquet"],["https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/clade_name_ref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/gene_family_ref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_cpm_gene_family_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_cpm_stratified_gene_family_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_cpm_stratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_cpm_unstratified_gene_family.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_cpm_unstratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_cpm_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_gene_family_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_relab_gene_family_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_relab_stratified_gene_family_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_relab_stratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_relab_unstratified_gene_family.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_relab_unstratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_relab_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_stratified_gene_family_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_stratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_unstratified_gene_family.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_unstratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genefamilies_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/genome_name_ref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/marker_abundance_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/marker_abundance_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/marker_presence_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/marker_presence_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_cpm_pathway_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_cpm_stratified_pathway_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_cpm_stratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_cpm_unstratified_pathway.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_cpm_unstratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_cpm_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_pathway_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_relab_pathway_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_relab_stratified_pathway_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_relab_stratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_relab_unstratified_pathway.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_relab_unstratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_relab_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_stratified_pathway_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_stratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_unstratified_pathway.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_unstratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathabundance_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathcoverage_pathway_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathcoverage_stratified_pathway_uniref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathcoverage_stratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathcoverage_unstratified_pathway.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathcoverage_unstratified_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathcoverage_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/pathway_ref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/relative_abundance_clade_name_species.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/relative_abundance_uuid.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/uniref_marker_ref.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/viral_clusters_genome_name.parquet","https://huggingface.co/datasets/waldronlab/metagenomics_mac/resolve/main/viral_clusters_uuid.parquet"],["reference","reference","genefamilies_cpm","genefamilies_cpm_stratified","genefamilies_cpm_stratified","genefamilies_cpm_unstratified","genefamilies_cpm_unstratified","genefamilies_cpm","genefamilies","genefamilies_relab","genefamilies_relab_stratified","genefamilies_relab_stratified","genefamilies_relab_unstratified","genefamilies_relab_unstratified","genefamilies_relab","genefamilies_stratified","genefamilies_stratified","genefamilies_unstratified","genefamilies_unstratified","genefamilies","reference","marker_abundance","marker_abundance","marker_presence","marker_presence","pathabundance_cpm","pathabundance_cpm_stratified","pathabundance_cpm_stratified","pathabundance_cpm_unstratified","pathabundance_cpm_unstratified","pathabundance_cpm","pathabundance","pathabundance_relab","pathabundance_relab_stratified","pathabundance_relab_stratified","pathabundance_relab_unstratified","pathabundance_relab_unstratified","pathabundance_relab","pathabundance_stratified","pathabundance_stratified","pathabundance_unstratified","pathabundance_unstratified","pathabundance","pathcoverage","pathcoverage_stratified","pathcoverage_stratified","pathcoverage_unstratified","pathcoverage_unstratified","pathcoverage","reference","relative_abundance","relative_abundance","reference","viral_clusters","viral_clusters"],[null,null,"HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN",null,"MetaPhlAn","MetaPhlAn","MetaPhlAn","MetaPhlAn","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN","HUMAnN",null,"MetaPhlAn","MetaPhlAn",null,"MetaPhlAn/Custom","MetaPhlAn/Custom"],["Reference file reporting all unique values of non-UUID identifiers in a parquet file.","Reference file reporting all unique values of non-UUID identifiers in a parquet file.","Gene family abundances normalized to Copies Per Million. This accounts for sequencing depth, making samples more comparable.","Gene family abundances (in CPM) that are taxonomically stratified, showing the contribution of each species to the total abundance of each gene family.","Gene family abundances (in CPM) that are taxonomically stratified, showing the contribution of each species to the total abundance of each gene family.","Total community-level gene family abundances (in CPM), without taxonomic stratification. This is the sum of the stratified values for each gene family.","Total community-level gene family abundances (in CPM), without taxonomic stratification. This is the sum of the stratified values for each gene family.","Gene family abundances normalized to Copies Per Million. This accounts for sequencing depth, making samples more comparable.","Abundance of gene families, typically identified by UniRef90 IDs. This is the raw, unnormalized output.","Gene family abundances converted to relative abundance. The abundances in each sample are scaled to sum to 100%.","Taxonomically stratified gene family abundances, expressed as relative abundance within each sample.","Taxonomically stratified gene family abundances, expressed as relative abundance within each sample.","Total community-level gene family abundances, expressed as relative abundance.","Total community-level gene family abundances, expressed as relative abundance.","Gene family abundances converted to relative abundance. The abundances in each sample are scaled to sum to 100%.","Raw gene family abundances (in RPK) that are taxonomically stratified, showing the contribution of each species.","Raw gene family abundances (in RPK) that are taxonomically stratified, showing the contribution of each species.","Total community-level gene family abundances (in RPK), without taxonomic stratification. This is equivalent to the main 'genefamilies' file.","Total community-level gene family abundances (in RPK), without taxonomic stratification. This is equivalent to the main 'genefamilies' file.","Abundance of gene families, typically identified by UniRef90 IDs. This is the raw, unnormalized output.","Reference file reporting all unique values of non-UUID identifiers in a parquet file.","Abundance of clade-specific marker genes. This is an intermediate file used by MetaPhlAn to calculate taxonomic relative abundances.","Abundance of clade-specific marker genes. This is an intermediate file used by MetaPhlAn to calculate taxonomic relative abundances.","A binary table indicating the presence (1) or absence (0) of specific marker genes for each taxon in a sample.","A binary table indicating the presence (1) or absence (0) of specific marker genes for each taxon in a sample.","Pathway abundances normalized to Copies Per Million to account for sequencing depth.","Pathway abundances (in CPM) that are taxonomically stratified, showing the contribution of each species.","Pathway abundances (in CPM) that are taxonomically stratified, showing the contribution of each species.","Total community-level pathway abundances (in CPM), without taxonomic stratification.","Total community-level pathway abundances (in CPM), without taxonomic stratification.","Pathway abundances normalized to Copies Per Million to account for sequencing depth.","Abundance of metabolic pathways (e.g., MetaCyc pathways). This is the raw, unnormalized output.","Pathway abundances converted to relative abundance. The abundances in each sample are scaled to sum to 100%.","Taxonomically stratified pathway abundances, expressed as relative abundance.","Taxonomically stratified pathway abundances, expressed as relative abundance.","Total community-level pathway abundances, expressed as relative abundance.","Total community-level pathway abundances, expressed as relative abundance.","Pathway abundances converted to relative abundance. The abundances in each sample are scaled to sum to 100%.","Raw pathway abundances (in RPK) that are taxonomically stratified, showing the contribution of each species.","Raw pathway abundances (in RPK) that are taxonomically stratified, showing the contribution of each species.","Total community-level pathway abundances (in RPK). This is equivalent to the main 'pathabundance' file.","Total community-level pathway abundances (in RPK). This is equivalent to the main 'pathabundance' file.","Abundance of metabolic pathways (e.g., MetaCyc pathways). This is the raw, unnormalized output.","The proportion of genes within a pathway that were detected in the sample. A value of 1.0 means all genes in the pathway were found.","Taxonomically stratified pathway coverage, showing the coverage of a pathway within the genome of a specific contributing species.","Taxonomically stratified pathway coverage, showing the coverage of a pathway within the genome of a specific contributing species.","Total community-level pathway coverage. This is equivalent to the main 'pathcoverage' file.","Total community-level pathway coverage. This is equivalent to the main 'pathcoverage' file.","The proportion of genes within a pathway that were detected in the sample. A value of 1.0 means all genes in the pathway were found.","Reference file reporting all unique values of non-UUID identifiers in a parquet file.","The primary output for taxonomic profiling, showing the relative abundance of each microbial taxon (from kingdom to species and strain level).","The primary output for taxonomic profiling, showing the relative abundance of each microbial taxon (from kingdom to species and strain level).","Reference file reporting all unique values of non-UUID identifiers in a parquet file.","Represents clusters of viral sequences, often used for viral strain or species-level analysis. The values typically represent the abundance of these viral clusters.","Represents clusters of viral sequences, often used for viral strain or species-level analysis. The values typically represent the abundance of these viral clusters."],[null,null,"Copies Per Million (CPM)","Copies Per Million (CPM)","Copies Per Million (CPM)","Copies Per Million (CPM)","Copies Per Million (CPM)","Copies Per Million (CPM)","Reads Per Kilobase (RPK)","Relative Abundance (%)","Relative Abundance (%)","Relative Abundance (%)","Relative Abundance (%)","Relative Abundance (%)","Relative Abundance (%)","Reads Per Kilobase (RPK)","Reads Per Kilobase (RPK)","Reads Per Kilobase (RPK)","Reads Per Kilobase (RPK)","Reads Per Kilobase (RPK)",null,"Mean coverage of marker genes","Mean coverage of marker genes","Binary (0 or 1)","Binary (0 or 1)","Copies Per Million (CPM)","Copies Per Million (CPM)","Copies Per Million (CPM)","Copies Per Million (CPM)","Copies Per Million (CPM)","Copies Per Million (CPM)","Reads Per Kilobase (RPK)","Relative Abundance (%)","Relative Abundance (%)","Relative Abundance (%)","Relative Abundance (%)","Relative Abundance (%)","Relative Abundance (%)","Reads Per Kilobase (RPK)","Reads Per Kilobase (RPK)","Reads Per Kilobase (RPK)","Reads Per Kilobase (RPK)","Reads Per Kilobase (RPK)","Proportion (0.0 to 1.0)","Proportion (0.0 to 1.0)","Proportion (0.0 to 1.0)","Proportion (0.0 to 1.0)","Proportion (0.0 to 1.0)","Proportion (0.0 to 1.0)",null,"Relative Abundance (%)","Relative Abundance (%)",null,"Varies (often Relative Abundance or CPM)","Varies (often Relative Abundance or CPM)"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>filename<\/th>\n      <th>url<\/th>\n      <th>data_type<\/th>\n      <th>tool<\/th>\n      <th>description<\/th>\n      <th>units_normalization<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"filename","targets":1},{"name":"url","targets":2},{"name":"data_type","targets":3},{"name":"tool","targets":4},{"name":"description","targets":5},{"name":"units_normalization","targets":6}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="sample-table">Sample Table<a class="anchor" aria-label="anchor" href="#sample-table"></a>
</h3>
<p>We can then examine the available samples and optionally select a
subset to query. This is done by browsing the
<code>sampleMetadata</code> object. Here is an example of how a set of
data might be selected for a meta-analysis with some sampling
parameters. Alternatively, this step can be left off altogether. If the
sample table is not provided when calling <code><a href="../reference/returnSamples.html">returnSamples()</a></code>,
data for all samples will simply be returned.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_table</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">study_name</span> <span class="op">==</span> <span class="st">"ZhangM_2023"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">select</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-36aa3d2a04d42bbc2145" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-36aa3d2a04d42bbc2145">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24"],["ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023"],["PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875"],["SAMN35813273","SAMN35813274","SAMN35813275","SAMN35813276","SAMN35813277","SAMN35813278","SAMN35813279","SAMN35813280","SAMN35813281","SAMN35813282","SAMN35813283","SAMN35813284","SAMN35813285","SAMN35813286","SAMN35813287","SAMN35813288","SAMN35813289","SAMN35813290","SAMN35813291","SAMN35813292","SAMN35813293","SAMN35813294","SAMN35813295","SAMN35813296"],["SRR25109817","SRR25109816","SRR25109805","SRR25109800","SRR25109799","SRR25109798","SRR25109797","SRR25109796","SRR25109795","SRR25109794","SRR25109815","SRR25109814","SRR25109813","SRR25109812","SRR25109811","SRR25109810","SRR25109809","SRR25109808","SRR25109807","SRR25109806","SRR25109804","SRR25109803","SRR25109802","SRR25109801"],["0807eb2a-a15e-4647-8e19-2600d8fda378","e0fbb54f-0249-4917-a4d7-bd68acb89c62","25172837-2849-4db3-be91-d54d6a815d00","39ddb5e7-97f6-4d3c-812b-9653b03f99b3","7b152a7d-e244-4e2b-b924-7195c7ecfb10","dd30f93b-7999-47a4-93fb-21971b899939","1406666f-04a8-43c9-983b-4ed62fd6da4a","fe3de3ca-3a14-4bd8-ae1c-0dad69edc9cd","8707e374-5ddb-4220-8cbf-364b8b0e7be1","22848a9c-66a6-4993-9058-cb6464edb42f","08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a","9baef0b2-93d2-4a40-8082-d357c7f8156a","09a9303d-d87d-4556-9672-04cbbcaf3d37","ac9f3532-90d8-412c-9c80-491037f0bcc2","eda61949-02dc-40ae-8dbe-bea2add85a52","1f007260-be6c-4a21-800a-ad9c36129a0d","e47a59bb-443a-405f-9c5d-02659d80e9e5","b3eaf3ab-43ef-4830-ab6d-12bafed3c61e","28f7352f-fe23-4003-93e1-41f4fedc6232","b07e2362-5851-4181-ba9a-15d9109ee4dd","677be4e3-722b-4e43-bd5a-36d8fbed6f86","0c817272-f873-475f-a401-dfe46a679a9f","7a3945d9-21bb-434a-9a4e-bfcdeb6194de","56aa2ad5-007d-407c-a644-48aac1e9a8f0"],["ZhangM_2023:F_PD1","ZhangM_2023:F_PD2","ZhangM_2023:F_PD3","ZhangM_2023:F_PD5","ZhangM_2023:F_PD7","ZhangM_2023:F_PD16","ZhangM_2023:F_PD19","ZhangM_2023:F_PD24","ZhangM_2023:F_PD27","ZhangM_2023:F_PD28","ZhangM_2023:F_PD29","ZhangM_2023:M_PD4","ZhangM_2023:M_PD6","ZhangM_2023:M_PD9","ZhangM_2023:M_PD10","ZhangM_2023:M_PD11","ZhangM_2023:M_PD13","ZhangM_2023:M_PD17","ZhangM_2023:M_PD18","ZhangM_2023:M_PD20","ZhangM_2023:M_PD21","ZhangM_2023:M_PD30","ZhangM_2023:M_PD31","ZhangM_2023:M_PD32"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease"],["NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845"],["feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces"],["UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988"],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606"],["Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case"],["NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152"],["Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease"],["NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845"],["Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long"],["SRR25109817","SRR25109816","SRR25109805","SRR25109800","SRR25109799","SRR25109798","SRR25109797","SRR25109796","SRR25109795","SRR25109794","SRR25109815","SRR25109814","SRR25109813","SRR25109812","SRR25109811","SRR25109810","SRR25109809","SRR25109808","SRR25109807","SRR25109806","SRR25109804","SRR25109803","SRR25109802","SRR25109801"],["WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS"],["281","274","282","277","280","287","291","289","290","292","291","280","280","290","282","286","286","290","289","290","290","293","292","291"],["5126812288","4637651878","4260703402","5073414290","4328969224","7837329037","5290309611","7625510674","8090442929","5777147435","5661210166","4332986236","5371153114","5654310289","4128355169","7327386738","7324599259","7868775948","7628779061","8040835028","5194000455","5795748371","5862491101","5693155117"],["PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875"],["SAMN35813273","SAMN35813274","SAMN35813275","SAMN35813276","SAMN35813277","SAMN35813278","SAMN35813279","SAMN35813280","SAMN35813281","SAMN35813282","SAMN35813283","SAMN35813284","SAMN35813285","SAMN35813286","SAMN35813287","SAMN35813288","SAMN35813289","SAMN35813290","SAMN35813291","SAMN35813292","SAMN35813293","SAMN35813294","SAMN35813295","SAMN35813296"],["1709080646","1540916207","1430374561","1688915874","1446219669","2588558077","1801807568","2520270145","2703434143","1922688037","1877184419","1448195135","1798553487","1915323502","1378503453","2419592743","2407455235","2593857558","2548175809","2672160392","1762093191","1933525398","1935544819","1895741776"],["THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY"],["public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public"],["fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra"],["gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3"],["gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1"],["SRX20862822","SRX20862823","SRX20862834","SRX20862839","SRX20862840","SRX20862841","SRX20862842","SRX20862843","SRX20862844","SRX20862845","SRX20862824","SRX20862825","SRX20862826","SRX20862827","SRX20862828","SRX20862829","SRX20862830","SRX20862831","SRX20862832","SRX20862833","SRX20862835","SRX20862836","SRX20862837","SRX20862838"],["China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China"],["Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia"],["China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou"],["Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE"],["RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM"],["METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC"],["human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome"],["ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA"],["2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z"],["2023-07-02T04:11:00Z","2023-07-02T04:11:00Z","2023-07-02T04:11:00Z","2023-07-02T04:10:00Z","2023-07-02T04:13:00Z","2023-07-02T04:18:00Z","2023-07-02T04:10:00Z","2023-07-02T04:12:00Z","2023-07-02T04:14:00Z","2023-07-02T04:13:00Z","2023-07-02T04:12:00Z","2023-07-02T04:11:00Z","2023-07-02T04:11:00Z","2023-07-02T04:14:00Z","2023-07-02T04:10:00Z","2023-07-02T04:12:00Z","2023-07-02T04:14:00Z","2023-07-02T04:18:00Z","2023-07-02T04:20:00Z","2023-07-02T04:16:00Z","2023-07-02T04:10:00Z","2023-07-02T04:11:00Z","2023-07-02T04:12:00Z","2023-07-02T04:12:00Z"],["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959"],["2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12"],["Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental"],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>study_name<\/th>\n      <th>BioProject<\/th>\n      <th>BioSample<\/th>\n      <th>NCBI_accession<\/th>\n      <th>uuid<\/th>\n      <th>curation_id<\/th>\n      <th>sample_id<\/th>\n      <th>subject_id<\/th>\n      <th>target_condition<\/th>\n      <th>target_condition_ontology_term_id<\/th>\n      <th>body_site<\/th>\n      <th>body_site_ontology_term_id<\/th>\n      <th>host_species<\/th>\n      <th>host_species_ontology_term_id<\/th>\n      <th>control<\/th>\n      <th>control_ontology_term_id<\/th>\n      <th>disease<\/th>\n      <th>disease_ontology_term_id<\/th>\n      <th>curator<\/th>\n      <th>uncurated_Run<\/th>\n      <th>uncurated_Assay.Type<\/th>\n      <th>uncurated_AvgSpotLen<\/th>\n      <th>uncurated_Bases<\/th>\n      <th>uncurated_BioProject<\/th>\n      <th>uncurated_BioSample<\/th>\n      <th>uncurated_Bytes<\/th>\n      <th>uncurated_Center.Name<\/th>\n      <th>uncurated_Consent<\/th>\n      <th>uncurated_DATASTORE.filetype<\/th>\n      <th>uncurated_DATASTORE.provider<\/th>\n      <th>uncurated_DATASTORE.region<\/th>\n      <th>uncurated_Experiment<\/th>\n      <th>uncurated_geo_loc_name_country<\/th>\n      <th>uncurated_geo_loc_name_country_continent<\/th>\n      <th>uncurated_geo_loc_name<\/th>\n      <th>uncurated_Instrument<\/th>\n      <th>uncurated_Library.Name<\/th>\n      <th>uncurated_LibraryLayout<\/th>\n      <th>uncurated_LibrarySelection<\/th>\n      <th>uncurated_LibrarySource<\/th>\n      <th>uncurated_Organism<\/th>\n      <th>uncurated_Platform<\/th>\n      <th>uncurated_ReleaseDate<\/th>\n      <th>uncurated_create_date<\/th>\n      <th>uncurated_version<\/th>\n      <th>uncurated_Sample.Name<\/th>\n      <th>uncurated_SRA.Study<\/th>\n      <th>uncurated_Collection_Date<\/th>\n      <th>uncurated_BioSampleModel<\/th>\n      <th>uncurated_HOST<\/th>\n      <th>uncurated_lat_lon<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"study_name","targets":1},{"name":"BioProject","targets":2},{"name":"BioSample","targets":3},{"name":"NCBI_accession","targets":4},{"name":"uuid","targets":5},{"name":"curation_id","targets":6},{"name":"sample_id","targets":7},{"name":"subject_id","targets":8},{"name":"target_condition","targets":9},{"name":"target_condition_ontology_term_id","targets":10},{"name":"body_site","targets":11},{"name":"body_site_ontology_term_id","targets":12},{"name":"host_species","targets":13},{"name":"host_species_ontology_term_id","targets":14},{"name":"control","targets":15},{"name":"control_ontology_term_id","targets":16},{"name":"disease","targets":17},{"name":"disease_ontology_term_id","targets":18},{"name":"curator","targets":19},{"name":"uncurated_Run","targets":20},{"name":"uncurated_Assay.Type","targets":21},{"name":"uncurated_AvgSpotLen","targets":22},{"name":"uncurated_Bases","targets":23},{"name":"uncurated_BioProject","targets":24},{"name":"uncurated_BioSample","targets":25},{"name":"uncurated_Bytes","targets":26},{"name":"uncurated_Center.Name","targets":27},{"name":"uncurated_Consent","targets":28},{"name":"uncurated_DATASTORE.filetype","targets":29},{"name":"uncurated_DATASTORE.provider","targets":30},{"name":"uncurated_DATASTORE.region","targets":31},{"name":"uncurated_Experiment","targets":32},{"name":"uncurated_geo_loc_name_country","targets":33},{"name":"uncurated_geo_loc_name_country_continent","targets":34},{"name":"uncurated_geo_loc_name","targets":35},{"name":"uncurated_Instrument","targets":36},{"name":"uncurated_Library.Name","targets":37},{"name":"uncurated_LibraryLayout","targets":38},{"name":"uncurated_LibrarySelection","targets":39},{"name":"uncurated_LibrarySource","targets":40},{"name":"uncurated_Organism","targets":41},{"name":"uncurated_Platform","targets":42},{"name":"uncurated_ReleaseDate","targets":43},{"name":"uncurated_create_date","targets":44},{"name":"uncurated_version","targets":45},{"name":"uncurated_Sample.Name","targets":46},{"name":"uncurated_SRA.Study","targets":47},{"name":"uncurated_Collection_Date","targets":48},{"name":"uncurated_BioSampleModel","targets":49},{"name":"uncurated_HOST","targets":50},{"name":"uncurated_lat_lon","targets":51}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="feature-table">Feature Table<a class="anchor" aria-label="anchor" href="#feature-table"></a>
</h3>
<p>For the feature table, we first have to find the reference file that
goes with the data type we are interested in. To do this, call
<code><a href="../reference/get_ref_info.html">get_ref_info()</a></code>. Search the ‘general_data_type’ column for
your data type and identify the reference file of interest. For a more
granular view of what exactly will be in each file, you can call
<code><a href="../reference/parquet_colinfo.html">parquet_colinfo()</a></code> with your data type and see which columns
will be included. Here, we can see that “clade_name_ref” is associated
with “relative_abundance”.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_ref_info.html">get_ref_info</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-febe03efa1a2d8d52a86" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-febe03efa1a2d8d52a86">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5"],["clade_name_ref","gene_family_ref","genome_name_ref","pathway_ref","uniref_marker_ref"],["relative_abundance","genefamilies","viral_clusters","pathabundance;pathcoverage","marker_abundance;marker_presence"],["MetaPhlAn","HUMAnN","MetaPhlAn","HUMAnN","MetaPhlAn"],["All unique values of clade_name and NCBI_tax_id found in the MetaPhlAn relative_abundance files in the same repo","All unique values of gene_family found in the HUMAnN genefamilies files in the same repo","All unique values of genome_name found in the MetaPhlAn viral_clusters files in the same repo","All unique values of pathway found in the HUMAnN pathabundance and pathcoverage files in the same repo","All unique values of uniref marker found in the MetaPhlAn marker_abundance and marker_presence files in the same repo"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>ref_file<\/th>\n      <th>general_data_type<\/th>\n      <th>tool<\/th>\n      <th>description<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"ref_file","targets":1},{"name":"general_data_type","targets":2},{"name":"tool","targets":3},{"name":"description","targets":4}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script><p>And here we can see exactly which columns are in
“clade_name_ref”.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/parquet_colinfo.html">parquet_colinfo</a></span><span class="op">(</span><span class="st">"relative_abundance"</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-1fb4450895fe099f74a1" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-1fb4450895fe099f74a1">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26"],["relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance","relative_abundance"],["clade_name","clade_name_kingdom","clade_name_phylum","clade_name_class","clade_name_order","clade_name_family","clade_name_genus","clade_name_species","clade_name_terminal","NCBI_tax_id","NCBI_tax_id_kingdom","NCBI_tax_id_phylum","NCBI_tax_id_class","NCBI_tax_id_order","NCBI_tax_id_family","NCBI_tax_id_genus","NCBI_tax_id_species","NCBI_tax_id_terminal","relative_abundance","additional_species","uuid","db_version","command","reads_processed","metaphlan_header","original_columns"],["character","character","character","character","character","character","character","character","character","character","character","character","character","character","character","character","character","character","float","character","character","character","character","character","character","character"],["The taxonomic lineage of the detected microbial clade","The taxonomic kingdom of the detected microbial clade","The taxonomic phylum of the detected microbial clade","The taxonomic class of the detected microbial clade","The taxonomic order of the detected microbial clade","The taxonomic family of the detected microbial clade","The taxonomic genus of the detected microbial clade","The taxonomic species of the detected microbial clade","The taxonomic terminal (strain, subspecies, etc.) of the detected microbial clade","The NCBI Taxonomy identifier for the clade in clade_name","The NCBI Taxonomy identifier for the kingdom in clade_name_kingdom","The NCBI Taxonomy identifier for the phylum in clade_name_phylum","The NCBI Taxonomy identifier for the class in clade_name_class","The NCBI Taxonomy identifier for the order in clade_name_order","The NCBI Taxonomy identifier for the family in clade_name_family","The NCBI Taxonomy identifier for the genus in clade_name_genus","The NCBI Taxonomy identifier for the species in clade_name_species","The NCBI Taxonomy identifier for the terminal (strain, subspecies, etc.) in clade_name_terminal","The proportion of the total microbial community represented by the clade","Any other species represented by the same set of detected markers","Sample UUID","MetaPhlAn database version(s) referenced","MetaPhlAn command given","Number of reads processed","MetaPhlAn's custom header row","Original MetaPhlAn column names"],["rname","rdata","rdata","rdata","rdata","rdata","rdata","rdata","rdata","rdata","rdata","rdata","rdata","rdata","rdata","rdata","rdata","rdata","assay","rdata","cname","cdata","cdata","cdata","cdata","cdata"],["clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref","clade_name_ref",null,null,null,null,null,null,null,null]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>general_data_type<\/th>\n      <th>col_name<\/th>\n      <th>col_class<\/th>\n      <th>description<\/th>\n      <th>se_role<\/th>\n      <th>ref_file<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"general_data_type","targets":1},{"name":"col_name","targets":2},{"name":"col_class","targets":3},{"name":"description","targets":4},{"name":"se_role","targets":5},{"name":"ref_file","targets":6}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script><p>Once we have found the correct reference file, we can filter it to
only contain features we are interested in. Here, we are interested in
values for all bacteria in the genus “Faecalibacterium”. We load the
reference file with <code><a href="../reference/load_ref.html">load_ref()</a></code>, and use a basic
<code><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl()</a></code> filtering method.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clade_name_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_ref.html">load_ref</a></span><span class="op">(</span><span class="st">"clade_name_ref"</span><span class="op">)</span></span>
<span><span class="va">feature_table</span> <span class="op">&lt;-</span> <span class="va">clade_name_ref</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Faecalibacterium"</span>, <span class="va">clade_name_genus</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-10b3b7155e8045a1b2ad" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-10b3b7155e8045a1b2ad">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18"],["k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_SGB15346","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_SGB15346|t__SGB15346","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15316","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15317","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15318","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15322","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15323","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15332","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15339","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15342","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_An122","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_An122|t__SGB15312","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_CLA_AA_H233","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_CLA_AA_H233|t__SGB15315","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_HTFF","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_HTFF|t__SGB15340"],["k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria"],["p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes"],["c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia"],["o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales"],["f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae"],["g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium"],[null,"s__Faecalibacterium_SGB15346","s__Faecalibacterium_SGB15346","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_sp_An122","s__Faecalibacterium_sp_An122","s__Faecalibacterium_sp_CLA_AA_H233","s__Faecalibacterium_sp_CLA_AA_H233","s__Faecalibacterium_sp_HTFF","s__Faecalibacterium_sp_HTFF"],[null,null,"t__SGB15346",null,"t__SGB15316","t__SGB15317","t__SGB15318","t__SGB15322","t__SGB15323","t__SGB15332","t__SGB15339","t__SGB15342",null,"t__SGB15312",null,"t__SGB15315",null,"t__SGB15340"],["2|1239|186801|186802|216572|216851","2|1239|186801|186802|216572|216851|","2|1239|186801|186802|216572|216851||","2|1239|186801|186802|216572|216851|853","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|1965551","2|1239|186801|186802|216572|216851|1965551|","2|1239|186801|186802|216572|216851|2881266","2|1239|186801|186802|216572|216851|2881266|","2|1239|186801|186802|216572|216851|2929491","2|1239|186801|186802|216572|216851|2929491|"],["2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2"],["1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239"],["186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801"],["186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802"],["216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572"],["216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851"],[null,"","","853","853","853","853","853","853","853","853","853","1965551","1965551","2881266","2881266","2929491","2929491"],[null,null,"",null,"","","","","","","","",null,"",null,"",null,""]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>clade_name<\/th>\n      <th>clade_name_kingdom<\/th>\n      <th>clade_name_phylum<\/th>\n      <th>clade_name_class<\/th>\n      <th>clade_name_order<\/th>\n      <th>clade_name_family<\/th>\n      <th>clade_name_genus<\/th>\n      <th>clade_name_species<\/th>\n      <th>clade_name_terminal<\/th>\n      <th>NCBI_tax_id<\/th>\n      <th>NCBI_tax_id_kingdom<\/th>\n      <th>NCBI_tax_id_phylum<\/th>\n      <th>NCBI_tax_id_class<\/th>\n      <th>NCBI_tax_id_order<\/th>\n      <th>NCBI_tax_id_family<\/th>\n      <th>NCBI_tax_id_genus<\/th>\n      <th>NCBI_tax_id_species<\/th>\n      <th>NCBI_tax_id_terminal<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"clade_name","targets":1},{"name":"clade_name_kingdom","targets":2},{"name":"clade_name_phylum","targets":3},{"name":"clade_name_class","targets":4},{"name":"clade_name_order","targets":5},{"name":"clade_name_family","targets":6},{"name":"clade_name_genus","targets":7},{"name":"clade_name_species","targets":8},{"name":"clade_name_terminal","targets":9},{"name":"NCBI_tax_id","targets":10},{"name":"NCBI_tax_id_kingdom","targets":11},{"name":"NCBI_tax_id_phylum","targets":12},{"name":"NCBI_tax_id_class","targets":13},{"name":"NCBI_tax_id_order","targets":14},{"name":"NCBI_tax_id_family","targets":15},{"name":"NCBI_tax_id_genus","targets":16},{"name":"NCBI_tax_id_species","targets":17},{"name":"NCBI_tax_id_terminal","targets":18}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script><p>As an extra consideration, MetaPhlAn relative abundance output
includes aggregate values for each taxonomic level. We can see above
that the first row returned has a value of <code>NA</code> for the
“clade_name_species” column, and a few other rows have <code>NA</code>
in the “clade_name_terminal” column. You may want to remove these rows
based on your analysis. To do so, we would simply re-filter the
file:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_table</span> <span class="op">&lt;-</span> <span class="va">clade_name_ref</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Faecalibacterium"</span>, <span class="va">clade_name_genus</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">clade_name_species</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">clade_name_terminal</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-4018eef1a407a0df6b52" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-4018eef1a407a0df6b52">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12"],["k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_SGB15346|t__SGB15346","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15316","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15317","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15318","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15322","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15323","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15332","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15339","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15342","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_An122|t__SGB15312","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_CLA_AA_H233|t__SGB15315","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_HTFF|t__SGB15340"],["k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria"],["p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes"],["c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia"],["o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales"],["f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae"],["g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium"],["s__Faecalibacterium_SGB15346","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_sp_An122","s__Faecalibacterium_sp_CLA_AA_H233","s__Faecalibacterium_sp_HTFF"],["t__SGB15346","t__SGB15316","t__SGB15317","t__SGB15318","t__SGB15322","t__SGB15323","t__SGB15332","t__SGB15339","t__SGB15342","t__SGB15312","t__SGB15315","t__SGB15340"],["2|1239|186801|186802|216572|216851||","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|1965551|","2|1239|186801|186802|216572|216851|2881266|","2|1239|186801|186802|216572|216851|2929491|"],["2","2","2","2","2","2","2","2","2","2","2","2"],["1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239"],["186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801"],["186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802"],["216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572"],["216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851"],["","853","853","853","853","853","853","853","853","1965551","2881266","2929491"],["","","","","","","","","","","",""]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>clade_name<\/th>\n      <th>clade_name_kingdom<\/th>\n      <th>clade_name_phylum<\/th>\n      <th>clade_name_class<\/th>\n      <th>clade_name_order<\/th>\n      <th>clade_name_family<\/th>\n      <th>clade_name_genus<\/th>\n      <th>clade_name_species<\/th>\n      <th>clade_name_terminal<\/th>\n      <th>NCBI_tax_id<\/th>\n      <th>NCBI_tax_id_kingdom<\/th>\n      <th>NCBI_tax_id_phylum<\/th>\n      <th>NCBI_tax_id_class<\/th>\n      <th>NCBI_tax_id_order<\/th>\n      <th>NCBI_tax_id_family<\/th>\n      <th>NCBI_tax_id_genus<\/th>\n      <th>NCBI_tax_id_species<\/th>\n      <th>NCBI_tax_id_terminal<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"clade_name","targets":1},{"name":"clade_name_kingdom","targets":2},{"name":"clade_name_phylum","targets":3},{"name":"clade_name_class","targets":4},{"name":"clade_name_order","targets":5},{"name":"clade_name_family","targets":6},{"name":"clade_name_genus","targets":7},{"name":"clade_name_species","targets":8},{"name":"clade_name_terminal","targets":9},{"name":"NCBI_tax_id","targets":10},{"name":"NCBI_tax_id_kingdom","targets":11},{"name":"NCBI_tax_id_phylum","targets":12},{"name":"NCBI_tax_id_class","targets":13},{"name":"NCBI_tax_id_order","targets":14},{"name":"NCBI_tax_id_family","targets":15},{"name":"NCBI_tax_id_genus","targets":16},{"name":"NCBI_tax_id_species","targets":17},{"name":"NCBI_tax_id_terminal","targets":18}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="returnsamples">returnSamples()<a class="anchor" aria-label="anchor" href="#returnsamples"></a>
</h3>
<p>We are now ready to retrieve our data. We simply pass our arguments
into <code><a href="../reference/returnSamples.html">returnSamples()</a></code>. It may take a minute, depending on
the amount of data requested or your available resources.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/returnSamples.html">returnSamples</a></span><span class="op">(</span>data_type <span class="op">=</span> <span class="st">"relative_abundance"</span>,</span>
<span>                            sample_data <span class="op">=</span> <span class="va">sample_table</span>,</span>
<span>                            feature_data <span class="op">=</span> <span class="va">feature_table</span>,</span>
<span>                            repo <span class="op">=</span> <span class="st">"waldronlab/metagenomics_mac"</span>,</span>
<span>                            local_files <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                            include_empty_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            dry_run <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">experiment</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 9 24 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): relative_abundance</span></span>
<span><span class="co">#&gt; rownames(9):</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_SGB15346|t__SGB15346</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15316</span></span>
<span><span class="co">#&gt;   ...</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15339</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_CLA_AA_H233|t__SGB15315</span></span>
<span><span class="co">#&gt; rowData names(19): clade_name clade_name_kingdom ...</span></span>
<span><span class="co">#&gt;   NCBI_tax_id_terminal additional_species</span></span>
<span><span class="co">#&gt; colnames(24): eda61949-02dc-40ae-8dbe-bea2add85a52</span></span>
<span><span class="co">#&gt;   e47a59bb-443a-405f-9c5d-02659d80e9e5 ...</span></span>
<span><span class="co">#&gt;   09a9303d-d87d-4556-9672-04cbbcaf3d37</span></span>
<span><span class="co">#&gt;   677be4e3-722b-4e43-bd5a-36d8fbed6f86</span></span>
<span><span class="co">#&gt; colData names(524): uuid db_version ...</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_digestion_issue</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_constipation</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: NULL</span></span>
<span><span class="co">#&gt; rowTree: NULL</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
<p>If you are finding that this function is failing due to rate limiting
(HTTP 429 error), hanging, or simply taking longer than you would like,
you can download the files associated with the data type you are
interested in and supply their paths to the “local_files” argument in
lieu of specifying the “repo” argument. It would look something like
this:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">local_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/path/to/relative_abundance_clade_name_species.parquet"</span>,</span>
<span>                 <span class="st">"/path/to/relative_abundance_uuid.parquet"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/returnSamples.html">returnSamples</a></span><span class="op">(</span>data_type <span class="op">=</span> <span class="st">"relative_abundance"</span>,</span>
<span>                            sample_data <span class="op">=</span> <span class="va">sample_table</span>,</span>
<span>                            feature_data <span class="op">=</span> <span class="va">feature_table</span>,</span>
<span>                            repo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                            local_files <span class="op">=</span> <span class="va">local_files</span>,</span>
<span>                            include_empty_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            dry_run <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Finally, you can check exactly which query is being called on the raw
parquet data by passing “TRUE” to the “dry_run” argument. This returns a
<code>tbl_duckdb_connection</code> object that can be passed to
<code><a href="https://dplyr.tidyverse.org/reference/explain.html" class="external-link">dplyr::show_query()</a></code>. To demonstrate:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_only</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/returnSamples.html">returnSamples</a></span><span class="op">(</span>data_type <span class="op">=</span> <span class="st">"relative_abundance"</span>,</span>
<span>                            sample_data <span class="op">=</span> <span class="va">sample_table</span>,</span>
<span>                            feature_data <span class="op">=</span> <span class="va">feature_table</span>,</span>
<span>                            repo <span class="op">=</span> <span class="st">"waldronlab/metagenomics_mac"</span>,</span>
<span>                            local_files <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                            include_empty_samples <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                            dry_run <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/explain.html" class="external-link">show_query</a></span><span class="op">(</span><span class="va">query_only</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;SQL&gt;</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">SELECT</span> q01.*</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">FROM</span> (</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">SELECT</span> relative_abundance_clade_name_species.*</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">FROM</span> relative_abundance_clade_name_species</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">WHERE</span> (clade_name_species = 's__Faecalibacterium_SGB15346')</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">UNION ALL</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">SELECT</span> relative_abundance_clade_name_species.*</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">FROM</span> relative_abundance_clade_name_species</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">WHERE</span> (clade_name_species = 's__Faecalibacterium_prausnitzii')</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">UNION ALL</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">SELECT</span> relative_abundance_clade_name_species.*</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">FROM</span> relative_abundance_clade_name_species</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">WHERE</span> (clade_name_species = 's__Faecalibacterium_sp_An122')</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">UNION ALL</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">SELECT</span> relative_abundance_clade_name_species.*</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">FROM</span> relative_abundance_clade_name_species</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">WHERE</span> (clade_name_species = 's__Faecalibacterium_sp_CLA_AA_H233')</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">UNION ALL</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">SELECT</span> relative_abundance_clade_name_species.*</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">FROM</span> relative_abundance_clade_name_species</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">WHERE</span> (clade_name_species = 's__Faecalibacterium_sp_HTFF')</span></span>
<span><span class="co">#&gt; ) q01</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">WHERE</span></span></span>
<span><span class="co">#&gt;   (clade_name_kingdom = 'k__Bacteria')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (clade_name_phylum = 'p__Firmicutes')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (clade_name_class = 'c__Clostridia')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (clade_name_order = 'o__Eubacteriales')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (clade_name_family = 'f__Oscillospiraceae')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (clade_name_genus = 'g__Faecalibacterium')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (NCBI_tax_id_kingdom = '2')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (NCBI_tax_id_phylum = '1239')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (NCBI_tax_id_class = '186801')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (NCBI_tax_id_order = '186802')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (NCBI_tax_id_family = '216572')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (NCBI_tax_id_genus = '216851')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (NCBI_tax_id_terminal = '')<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (NCBI_tax_id IN ('2|1239|186801|186802|216572|216851||', '2|1239|186801|186802|216572|216851|853|', '2|1239|186801|186802|216572|216851|1965551|', '2|1239|186801|186802|216572|216851|2881266|', '2|1239|186801|186802|216572|216851|2929491|'))<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (NCBI_tax_id_species IN ('', '853', '1965551', '2881266', '2929491'))<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (clade_name IN ('k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_SGB15346|t__SGB15346', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15316', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15317', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15318', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15322', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15323', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15332', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15339', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15342', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_An122|t__SGB15312', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_CLA_AA_H233|t__SGB15315', 'k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_HTFF|t__SGB15340'))<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (clade_name_terminal IN ('t__SGB15346', 't__SGB15316', 't__SGB15317', 't__SGB15318', 't__SGB15322', 't__SGB15323', 't__SGB15332', 't__SGB15339', 't__SGB15342', 't__SGB15312', 't__SGB15315', 't__SGB15340'))<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (uuid IN ('0807eb2a-a15e-4647-8e19-2600d8fda378', 'e0fbb54f-0249-4917-a4d7-bd68acb89c62', '25172837-2849-4db3-be91-d54d6a815d00', '39ddb5e7-97f6-4d3c-812b-9653b03f99b3', '7b152a7d-e244-4e2b-b924-7195c7ecfb10', 'dd30f93b-7999-47a4-93fb-21971b899939', '1406666f-04a8-43c9-983b-4ed62fd6da4a', 'fe3de3ca-3a14-4bd8-ae1c-0dad69edc9cd', '8707e374-5ddb-4220-8cbf-364b8b0e7be1', '22848a9c-66a6-4993-9058-cb6464edb42f', '08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a', '9baef0b2-93d2-4a40-8082-d357c7f8156a', '09a9303d-d87d-4556-9672-04cbbcaf3d37', 'ac9f3532-90d8-412c-9c80-491037f0bcc2', 'eda61949-02dc-40ae-8dbe-bea2add85a52', '1f007260-be6c-4a21-800a-ad9c36129a0d', 'e47a59bb-443a-405f-9c5d-02659d80e9e5', 'b3eaf3ab-43ef-4830-ab6d-12bafed3c61e', '28f7352f-fe23-4003-93e1-41f4fedc6232', 'b07e2362-5851-4181-ba9a-15d9109ee4dd', '677be4e3-722b-4e43-bd5a-36d8fbed6f86', '0c817272-f873-475f-a401-dfe46a679a9f', '7a3945d9-21bb-434a-9a4e-bfcdeb6194de', '56aa2ad5-007d-407c-a644-48aac1e9a8f0'))</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="piecewise-workflow">Piecewise Workflow<a class="anchor" aria-label="anchor" href="#piecewise-workflow"></a>
</h2>
<p>In the piecewise workflow, we use the functions
<code>accessParquetData</code> and <code>loadParquetData</code>
separately to have a little more control over our data. First,
<code>accessParquetData</code> sets up a DuckDB connection to the
parquet files that are either stored locally or hosted in a remote repo
(in this case, Hugging Face). <code>loadParquetData</code> then takes an
argument for data type as well as any filters and loads the requested
data into R as a Tree Summarized Experiment. We can even use
<code>dplyr</code> functions to customize the SQL query.</p>
<div class="section level3">
<h3 id="file-selection-1">File Selection<a class="anchor" aria-label="anchor" href="#file-selection-1"></a>
</h3>
<p>We can use the same resources of <code><a href="../reference/get_repo_info.html">get_repo_info()</a></code> and
<code><a href="../reference/get_hf_parquet_urls.html">get_hf_parquet_urls()</a></code> to determine our files of interest
and <code>data_type</code> value. As before, we can download any files
and alternatively supply them as local files.</p>
<p>We then run <code>accessParquetData</code>, which creates the DuckDB
connection and sets up DuckDB “VIEW” objects for each available file,
whether remotely hosted or locally stored. It returns a DuckDB
connection object. If you are using a remote repo, the
<code>data_types</code> argument can be left blank to access all
available files, or a smaller number can be provided. DBI functions can
then be used to see which views are now available.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">con</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/accessParquetData.html">accessParquetData</a></span><span class="op">(</span>dbdir <span class="op">=</span> <span class="st">":memory:"</span>,</span>
<span>                         repo <span class="op">=</span> <span class="st">"waldronlab/metagenomics_mac"</span>,</span>
<span>                         local_files <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                         data_types <span class="op">=</span> <span class="st">"relative_abundance"</span><span class="op">)</span></span>
<span><span class="fu">DBI</span><span class="fu">::</span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbListTables.html" class="external-link">dbListTables</a></span><span class="op">(</span><span class="va">con</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "relative_abundance_clade_name_species"</span></span>
<span><span class="co">#&gt; [2] "relative_abundance_uuid"</span></span></code></pre></div>
<p>Supplying local files would look much as expected, and the
“data_types” argument need not be supplied:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">local_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/path/to/relative_abundance_clade_name_species.parquet"</span>,</span>
<span>                 <span class="st">"/path/to/relative_abundance_uuid.parquet"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">con</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/accessParquetData.html">accessParquetData</a></span><span class="op">(</span>dbdir <span class="op">=</span> <span class="st">":memory:"</span>,</span>
<span>                         repo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                         local_files <span class="op">=</span> <span class="va">local_files</span>,</span>
<span>                         data_types <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="selecting-samples">Selecting Samples<a class="anchor" aria-label="anchor" href="#selecting-samples"></a>
</h3>
<p>The sample selection process is the same as in the standard workflow,
with the exception that we are only passing the UUID column to the
loading function. Here we will create the same table as earlier, then
pull just the UUIDS. This is again an optional step, as you may want to
get data across all samples for meta-analyses.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_table</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">study_name</span> <span class="op">==</span> <span class="st">"ZhangM_2023"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">select</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-5b1b2f4ad92281566982" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-5b1b2f4ad92281566982">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24"],["ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023","ZhangM_2023"],["PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875"],["SAMN35813273","SAMN35813274","SAMN35813275","SAMN35813276","SAMN35813277","SAMN35813278","SAMN35813279","SAMN35813280","SAMN35813281","SAMN35813282","SAMN35813283","SAMN35813284","SAMN35813285","SAMN35813286","SAMN35813287","SAMN35813288","SAMN35813289","SAMN35813290","SAMN35813291","SAMN35813292","SAMN35813293","SAMN35813294","SAMN35813295","SAMN35813296"],["SRR25109817","SRR25109816","SRR25109805","SRR25109800","SRR25109799","SRR25109798","SRR25109797","SRR25109796","SRR25109795","SRR25109794","SRR25109815","SRR25109814","SRR25109813","SRR25109812","SRR25109811","SRR25109810","SRR25109809","SRR25109808","SRR25109807","SRR25109806","SRR25109804","SRR25109803","SRR25109802","SRR25109801"],["0807eb2a-a15e-4647-8e19-2600d8fda378","e0fbb54f-0249-4917-a4d7-bd68acb89c62","25172837-2849-4db3-be91-d54d6a815d00","39ddb5e7-97f6-4d3c-812b-9653b03f99b3","7b152a7d-e244-4e2b-b924-7195c7ecfb10","dd30f93b-7999-47a4-93fb-21971b899939","1406666f-04a8-43c9-983b-4ed62fd6da4a","fe3de3ca-3a14-4bd8-ae1c-0dad69edc9cd","8707e374-5ddb-4220-8cbf-364b8b0e7be1","22848a9c-66a6-4993-9058-cb6464edb42f","08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a","9baef0b2-93d2-4a40-8082-d357c7f8156a","09a9303d-d87d-4556-9672-04cbbcaf3d37","ac9f3532-90d8-412c-9c80-491037f0bcc2","eda61949-02dc-40ae-8dbe-bea2add85a52","1f007260-be6c-4a21-800a-ad9c36129a0d","e47a59bb-443a-405f-9c5d-02659d80e9e5","b3eaf3ab-43ef-4830-ab6d-12bafed3c61e","28f7352f-fe23-4003-93e1-41f4fedc6232","b07e2362-5851-4181-ba9a-15d9109ee4dd","677be4e3-722b-4e43-bd5a-36d8fbed6f86","0c817272-f873-475f-a401-dfe46a679a9f","7a3945d9-21bb-434a-9a4e-bfcdeb6194de","56aa2ad5-007d-407c-a644-48aac1e9a8f0"],["ZhangM_2023:F_PD1","ZhangM_2023:F_PD2","ZhangM_2023:F_PD3","ZhangM_2023:F_PD5","ZhangM_2023:F_PD7","ZhangM_2023:F_PD16","ZhangM_2023:F_PD19","ZhangM_2023:F_PD24","ZhangM_2023:F_PD27","ZhangM_2023:F_PD28","ZhangM_2023:F_PD29","ZhangM_2023:M_PD4","ZhangM_2023:M_PD6","ZhangM_2023:M_PD9","ZhangM_2023:M_PD10","ZhangM_2023:M_PD11","ZhangM_2023:M_PD13","ZhangM_2023:M_PD17","ZhangM_2023:M_PD18","ZhangM_2023:M_PD20","ZhangM_2023:M_PD21","ZhangM_2023:M_PD30","ZhangM_2023:M_PD31","ZhangM_2023:M_PD32"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease"],["NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845"],["feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces","feces"],["UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988","UBERON:0001988"],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606"],["Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Case"],["NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152","NCIT:C49152"],["Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease","Parkinson Disease"],["NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845","NCIT:C26845"],["Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long","Kaelyn Long"],["SRR25109817","SRR25109816","SRR25109805","SRR25109800","SRR25109799","SRR25109798","SRR25109797","SRR25109796","SRR25109795","SRR25109794","SRR25109815","SRR25109814","SRR25109813","SRR25109812","SRR25109811","SRR25109810","SRR25109809","SRR25109808","SRR25109807","SRR25109806","SRR25109804","SRR25109803","SRR25109802","SRR25109801"],["WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS","WGS"],["281","274","282","277","280","287","291","289","290","292","291","280","280","290","282","286","286","290","289","290","290","293","292","291"],["5126812288","4637651878","4260703402","5073414290","4328969224","7837329037","5290309611","7625510674","8090442929","5777147435","5661210166","4332986236","5371153114","5654310289","4128355169","7327386738","7324599259","7868775948","7628779061","8040835028","5194000455","5795748371","5862491101","5693155117"],["PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875","PRJNA985875"],["SAMN35813273","SAMN35813274","SAMN35813275","SAMN35813276","SAMN35813277","SAMN35813278","SAMN35813279","SAMN35813280","SAMN35813281","SAMN35813282","SAMN35813283","SAMN35813284","SAMN35813285","SAMN35813286","SAMN35813287","SAMN35813288","SAMN35813289","SAMN35813290","SAMN35813291","SAMN35813292","SAMN35813293","SAMN35813294","SAMN35813295","SAMN35813296"],["1709080646","1540916207","1430374561","1688915874","1446219669","2588558077","1801807568","2520270145","2703434143","1922688037","1877184419","1448195135","1798553487","1915323502","1378503453","2419592743","2407455235","2593857558","2548175809","2672160392","1762093191","1933525398","1935544819","1895741776"],["THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY","THE AFFILIATED HUAIAN NO.1 PEOPLE HOSPITAL OF NANJING MEDICAL UNIVERSITY"],["public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public","public"],["fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra","fastq,run.zq,sra"],["gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3","gs,ncbi,s3"],["gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1","gs.us-east1,ncbi.public,s3.us-east-1"],["SRX20862822","SRX20862823","SRX20862834","SRX20862839","SRX20862840","SRX20862841","SRX20862842","SRX20862843","SRX20862844","SRX20862845","SRX20862824","SRX20862825","SRX20862826","SRX20862827","SRX20862828","SRX20862829","SRX20862830","SRX20862831","SRX20862832","SRX20862833","SRX20862835","SRX20862836","SRX20862837","SRX20862838"],["China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China","China"],["Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia","Asia"],["China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou","China:HangZhou"],["Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000","Illumina NovaSeq 6000"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE","SINGLE"],["RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM","RANDOM"],["METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC","METAGENOMIC"],["human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome","human metagenome"],["ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA","ILLUMINA"],["2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z","2023-07-02T00:00:00Z"],["2023-07-02T04:11:00Z","2023-07-02T04:11:00Z","2023-07-02T04:11:00Z","2023-07-02T04:10:00Z","2023-07-02T04:13:00Z","2023-07-02T04:18:00Z","2023-07-02T04:10:00Z","2023-07-02T04:12:00Z","2023-07-02T04:14:00Z","2023-07-02T04:13:00Z","2023-07-02T04:12:00Z","2023-07-02T04:11:00Z","2023-07-02T04:11:00Z","2023-07-02T04:14:00Z","2023-07-02T04:10:00Z","2023-07-02T04:12:00Z","2023-07-02T04:14:00Z","2023-07-02T04:18:00Z","2023-07-02T04:20:00Z","2023-07-02T04:16:00Z","2023-07-02T04:10:00Z","2023-07-02T04:11:00Z","2023-07-02T04:12:00Z","2023-07-02T04:12:00Z"],["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],["F_PD1","F_PD2","F_PD3","F_PD5","F_PD7","F_PD16","F_PD19","F_PD24","F_PD27","F_PD28","F_PD29","M_PD4","M_PD6","M_PD9","M_PD10","M_PD11","M_PD13","M_PD17","M_PD18","M_PD20","M_PD21","M_PD30","M_PD31","M_PD32"],["SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959","SRP446959"],["2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12","2022-12"],["Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental"],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E","30 N 120 E"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>study_name<\/th>\n      <th>BioProject<\/th>\n      <th>BioSample<\/th>\n      <th>NCBI_accession<\/th>\n      <th>uuid<\/th>\n      <th>curation_id<\/th>\n      <th>sample_id<\/th>\n      <th>subject_id<\/th>\n      <th>target_condition<\/th>\n      <th>target_condition_ontology_term_id<\/th>\n      <th>body_site<\/th>\n      <th>body_site_ontology_term_id<\/th>\n      <th>host_species<\/th>\n      <th>host_species_ontology_term_id<\/th>\n      <th>control<\/th>\n      <th>control_ontology_term_id<\/th>\n      <th>disease<\/th>\n      <th>disease_ontology_term_id<\/th>\n      <th>curator<\/th>\n      <th>uncurated_Run<\/th>\n      <th>uncurated_Assay.Type<\/th>\n      <th>uncurated_AvgSpotLen<\/th>\n      <th>uncurated_Bases<\/th>\n      <th>uncurated_BioProject<\/th>\n      <th>uncurated_BioSample<\/th>\n      <th>uncurated_Bytes<\/th>\n      <th>uncurated_Center.Name<\/th>\n      <th>uncurated_Consent<\/th>\n      <th>uncurated_DATASTORE.filetype<\/th>\n      <th>uncurated_DATASTORE.provider<\/th>\n      <th>uncurated_DATASTORE.region<\/th>\n      <th>uncurated_Experiment<\/th>\n      <th>uncurated_geo_loc_name_country<\/th>\n      <th>uncurated_geo_loc_name_country_continent<\/th>\n      <th>uncurated_geo_loc_name<\/th>\n      <th>uncurated_Instrument<\/th>\n      <th>uncurated_Library.Name<\/th>\n      <th>uncurated_LibraryLayout<\/th>\n      <th>uncurated_LibrarySelection<\/th>\n      <th>uncurated_LibrarySource<\/th>\n      <th>uncurated_Organism<\/th>\n      <th>uncurated_Platform<\/th>\n      <th>uncurated_ReleaseDate<\/th>\n      <th>uncurated_create_date<\/th>\n      <th>uncurated_version<\/th>\n      <th>uncurated_Sample.Name<\/th>\n      <th>uncurated_SRA.Study<\/th>\n      <th>uncurated_Collection_Date<\/th>\n      <th>uncurated_BioSampleModel<\/th>\n      <th>uncurated_HOST<\/th>\n      <th>uncurated_lat_lon<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"study_name","targets":1},{"name":"BioProject","targets":2},{"name":"BioSample","targets":3},{"name":"NCBI_accession","targets":4},{"name":"uuid","targets":5},{"name":"curation_id","targets":6},{"name":"sample_id","targets":7},{"name":"subject_id","targets":8},{"name":"target_condition","targets":9},{"name":"target_condition_ontology_term_id","targets":10},{"name":"body_site","targets":11},{"name":"body_site_ontology_term_id","targets":12},{"name":"host_species","targets":13},{"name":"host_species_ontology_term_id","targets":14},{"name":"control","targets":15},{"name":"control_ontology_term_id","targets":16},{"name":"disease","targets":17},{"name":"disease_ontology_term_id","targets":18},{"name":"curator","targets":19},{"name":"uncurated_Run","targets":20},{"name":"uncurated_Assay.Type","targets":21},{"name":"uncurated_AvgSpotLen","targets":22},{"name":"uncurated_Bases","targets":23},{"name":"uncurated_BioProject","targets":24},{"name":"uncurated_BioSample","targets":25},{"name":"uncurated_Bytes","targets":26},{"name":"uncurated_Center.Name","targets":27},{"name":"uncurated_Consent","targets":28},{"name":"uncurated_DATASTORE.filetype","targets":29},{"name":"uncurated_DATASTORE.provider","targets":30},{"name":"uncurated_DATASTORE.region","targets":31},{"name":"uncurated_Experiment","targets":32},{"name":"uncurated_geo_loc_name_country","targets":33},{"name":"uncurated_geo_loc_name_country_continent","targets":34},{"name":"uncurated_geo_loc_name","targets":35},{"name":"uncurated_Instrument","targets":36},{"name":"uncurated_Library.Name","targets":37},{"name":"uncurated_LibraryLayout","targets":38},{"name":"uncurated_LibrarySelection","targets":39},{"name":"uncurated_LibrarySource","targets":40},{"name":"uncurated_Organism","targets":41},{"name":"uncurated_Platform","targets":42},{"name":"uncurated_ReleaseDate","targets":43},{"name":"uncurated_create_date","targets":44},{"name":"uncurated_version","targets":45},{"name":"uncurated_Sample.Name","targets":46},{"name":"uncurated_SRA.Study","targets":47},{"name":"uncurated_Collection_Date","targets":48},{"name":"uncurated_BioSampleModel","targets":49},{"name":"uncurated_HOST","targets":50},{"name":"uncurated_lat_lon","targets":51}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_uuids</span> <span class="op">&lt;-</span> <span class="va">sample_table</span><span class="op">$</span><span class="va">uuid</span></span>
<span><span class="va">selected_uuids</span></span>
<span><span class="co">#&gt;  [1] "0807eb2a-a15e-4647-8e19-2600d8fda378"</span></span>
<span><span class="co">#&gt;  [2] "e0fbb54f-0249-4917-a4d7-bd68acb89c62"</span></span>
<span><span class="co">#&gt;  [3] "25172837-2849-4db3-be91-d54d6a815d00"</span></span>
<span><span class="co">#&gt;  [4] "39ddb5e7-97f6-4d3c-812b-9653b03f99b3"</span></span>
<span><span class="co">#&gt;  [5] "7b152a7d-e244-4e2b-b924-7195c7ecfb10"</span></span>
<span><span class="co">#&gt;  [6] "dd30f93b-7999-47a4-93fb-21971b899939"</span></span>
<span><span class="co">#&gt;  [7] "1406666f-04a8-43c9-983b-4ed62fd6da4a"</span></span>
<span><span class="co">#&gt;  [8] "fe3de3ca-3a14-4bd8-ae1c-0dad69edc9cd"</span></span>
<span><span class="co">#&gt;  [9] "8707e374-5ddb-4220-8cbf-364b8b0e7be1"</span></span>
<span><span class="co">#&gt; [10] "22848a9c-66a6-4993-9058-cb6464edb42f"</span></span>
<span><span class="co">#&gt; [11] "08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a"</span></span>
<span><span class="co">#&gt; [12] "9baef0b2-93d2-4a40-8082-d357c7f8156a"</span></span>
<span><span class="co">#&gt; [13] "09a9303d-d87d-4556-9672-04cbbcaf3d37"</span></span>
<span><span class="co">#&gt; [14] "ac9f3532-90d8-412c-9c80-491037f0bcc2"</span></span>
<span><span class="co">#&gt; [15] "eda61949-02dc-40ae-8dbe-bea2add85a52"</span></span>
<span><span class="co">#&gt; [16] "1f007260-be6c-4a21-800a-ad9c36129a0d"</span></span>
<span><span class="co">#&gt; [17] "e47a59bb-443a-405f-9c5d-02659d80e9e5"</span></span>
<span><span class="co">#&gt; [18] "b3eaf3ab-43ef-4830-ab6d-12bafed3c61e"</span></span>
<span><span class="co">#&gt; [19] "28f7352f-fe23-4003-93e1-41f4fedc6232"</span></span>
<span><span class="co">#&gt; [20] "b07e2362-5851-4181-ba9a-15d9109ee4dd"</span></span>
<span><span class="co">#&gt; [21] "677be4e3-722b-4e43-bd5a-36d8fbed6f86"</span></span>
<span><span class="co">#&gt; [22] "0c817272-f873-475f-a401-dfe46a679a9f"</span></span>
<span><span class="co">#&gt; [23] "7a3945d9-21bb-434a-9a4e-bfcdeb6194de"</span></span>
<span><span class="co">#&gt; [24] "56aa2ad5-007d-407c-a644-48aac1e9a8f0"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="selecting-features">Selecting Features<a class="anchor" aria-label="anchor" href="#selecting-features"></a>
</h3>
<p>Our feature table is treated similarly, in that we are no longer
providing the whole table. Instead, we construct our filtering arguments
as a named list. The element name is the column to filter by, and the
element values will be exact matches. If this is our previous table:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_table</span> <span class="op">&lt;-</span> <span class="va">clade_name_ref</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Faecalibacterium"</span>, <span class="va">clade_name_genus</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">clade_name_species</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">clade_name_terminal</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-25c3e940e6859592f801" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-25c3e940e6859592f801">{"x":{"filter":"none","vertical":false,"extensions":["Responsive"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12"],["k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_SGB15346|t__SGB15346","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15316","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15317","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15318","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15322","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15323","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15332","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15339","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15342","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_An122|t__SGB15312","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_CLA_AA_H233|t__SGB15315","k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_HTFF|t__SGB15340"],["k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria","k__Bacteria"],["p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes"],["c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia","c__Clostridia"],["o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales","o__Eubacteriales"],["f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae","f__Oscillospiraceae"],["g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium","g__Faecalibacterium"],["s__Faecalibacterium_SGB15346","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_prausnitzii","s__Faecalibacterium_sp_An122","s__Faecalibacterium_sp_CLA_AA_H233","s__Faecalibacterium_sp_HTFF"],["t__SGB15346","t__SGB15316","t__SGB15317","t__SGB15318","t__SGB15322","t__SGB15323","t__SGB15332","t__SGB15339","t__SGB15342","t__SGB15312","t__SGB15315","t__SGB15340"],["2|1239|186801|186802|216572|216851||","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|853|","2|1239|186801|186802|216572|216851|1965551|","2|1239|186801|186802|216572|216851|2881266|","2|1239|186801|186802|216572|216851|2929491|"],["2","2","2","2","2","2","2","2","2","2","2","2"],["1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239","1239"],["186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801","186801"],["186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802","186802"],["216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572","216572"],["216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851","216851"],["","853","853","853","853","853","853","853","853","1965551","2881266","2929491"],["","","","","","","","","","","",""]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>clade_name<\/th>\n      <th>clade_name_kingdom<\/th>\n      <th>clade_name_phylum<\/th>\n      <th>clade_name_class<\/th>\n      <th>clade_name_order<\/th>\n      <th>clade_name_family<\/th>\n      <th>clade_name_genus<\/th>\n      <th>clade_name_species<\/th>\n      <th>clade_name_terminal<\/th>\n      <th>NCBI_tax_id<\/th>\n      <th>NCBI_tax_id_kingdom<\/th>\n      <th>NCBI_tax_id_phylum<\/th>\n      <th>NCBI_tax_id_class<\/th>\n      <th>NCBI_tax_id_order<\/th>\n      <th>NCBI_tax_id_family<\/th>\n      <th>NCBI_tax_id_genus<\/th>\n      <th>NCBI_tax_id_species<\/th>\n      <th>NCBI_tax_id_terminal<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"clade_name","targets":1},{"name":"clade_name_kingdom","targets":2},{"name":"clade_name_phylum","targets":3},{"name":"clade_name_class","targets":4},{"name":"clade_name_order","targets":5},{"name":"clade_name_family","targets":6},{"name":"clade_name_genus","targets":7},{"name":"clade_name_species","targets":8},{"name":"clade_name_terminal","targets":9},{"name":"NCBI_tax_id","targets":10},{"name":"NCBI_tax_id_kingdom","targets":11},{"name":"NCBI_tax_id_phylum","targets":12},{"name":"NCBI_tax_id_class","targets":13},{"name":"NCBI_tax_id_order","targets":14},{"name":"NCBI_tax_id_family","targets":15},{"name":"NCBI_tax_id_genus","targets":16},{"name":"NCBI_tax_id_species","targets":17},{"name":"NCBI_tax_id_terminal","targets":18}],"order":[],"autoWidth":false,"orderClasses":false,"responsive":true}},"evals":[],"jsHooks":[]}</script><p>We would set up the filters like so. These two columns are the
minimum required to produce the same result as our prior example.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filter_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clade_name_species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">feature_table</span><span class="op">$</span><span class="va">clade_name_species</span><span class="op">)</span>,</span>
<span>                      clade_name_terminal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">feature_table</span><span class="op">$</span><span class="va">clade_name_terminal</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">filter_values</span></span>
<span><span class="co">#&gt; $clade_name_species</span></span>
<span><span class="co">#&gt; [1] "s__Faecalibacterium_SGB15346"       "s__Faecalibacterium_prausnitzii"   </span></span>
<span><span class="co">#&gt; [3] "s__Faecalibacterium_sp_An122"       "s__Faecalibacterium_sp_CLA_AA_H233"</span></span>
<span><span class="co">#&gt; [5] "s__Faecalibacterium_sp_HTFF"       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $clade_name_terminal</span></span>
<span><span class="co">#&gt;  [1] "t__SGB15346" "t__SGB15316" "t__SGB15317" "t__SGB15318" "t__SGB15322"</span></span>
<span><span class="co">#&gt;  [6] "t__SGB15323" "t__SGB15332" "t__SGB15339" "t__SGB15342" "t__SGB15312"</span></span>
<span><span class="co">#&gt; [11] "t__SGB15315" "t__SGB15340"</span></span></code></pre></div>
<p>Here is also where we supply our UUIDs.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filter_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>uuid <span class="op">=</span> <span class="va">selected_uuids</span><span class="op">)</span>, <span class="va">filter_values</span><span class="op">)</span></span>
<span><span class="va">filter_values</span></span>
<span><span class="co">#&gt; $uuid</span></span>
<span><span class="co">#&gt;  [1] "0807eb2a-a15e-4647-8e19-2600d8fda378"</span></span>
<span><span class="co">#&gt;  [2] "e0fbb54f-0249-4917-a4d7-bd68acb89c62"</span></span>
<span><span class="co">#&gt;  [3] "25172837-2849-4db3-be91-d54d6a815d00"</span></span>
<span><span class="co">#&gt;  [4] "39ddb5e7-97f6-4d3c-812b-9653b03f99b3"</span></span>
<span><span class="co">#&gt;  [5] "7b152a7d-e244-4e2b-b924-7195c7ecfb10"</span></span>
<span><span class="co">#&gt;  [6] "dd30f93b-7999-47a4-93fb-21971b899939"</span></span>
<span><span class="co">#&gt;  [7] "1406666f-04a8-43c9-983b-4ed62fd6da4a"</span></span>
<span><span class="co">#&gt;  [8] "fe3de3ca-3a14-4bd8-ae1c-0dad69edc9cd"</span></span>
<span><span class="co">#&gt;  [9] "8707e374-5ddb-4220-8cbf-364b8b0e7be1"</span></span>
<span><span class="co">#&gt; [10] "22848a9c-66a6-4993-9058-cb6464edb42f"</span></span>
<span><span class="co">#&gt; [11] "08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a"</span></span>
<span><span class="co">#&gt; [12] "9baef0b2-93d2-4a40-8082-d357c7f8156a"</span></span>
<span><span class="co">#&gt; [13] "09a9303d-d87d-4556-9672-04cbbcaf3d37"</span></span>
<span><span class="co">#&gt; [14] "ac9f3532-90d8-412c-9c80-491037f0bcc2"</span></span>
<span><span class="co">#&gt; [15] "eda61949-02dc-40ae-8dbe-bea2add85a52"</span></span>
<span><span class="co">#&gt; [16] "1f007260-be6c-4a21-800a-ad9c36129a0d"</span></span>
<span><span class="co">#&gt; [17] "e47a59bb-443a-405f-9c5d-02659d80e9e5"</span></span>
<span><span class="co">#&gt; [18] "b3eaf3ab-43ef-4830-ab6d-12bafed3c61e"</span></span>
<span><span class="co">#&gt; [19] "28f7352f-fe23-4003-93e1-41f4fedc6232"</span></span>
<span><span class="co">#&gt; [20] "b07e2362-5851-4181-ba9a-15d9109ee4dd"</span></span>
<span><span class="co">#&gt; [21] "677be4e3-722b-4e43-bd5a-36d8fbed6f86"</span></span>
<span><span class="co">#&gt; [22] "0c817272-f873-475f-a401-dfe46a679a9f"</span></span>
<span><span class="co">#&gt; [23] "7a3945d9-21bb-434a-9a4e-bfcdeb6194de"</span></span>
<span><span class="co">#&gt; [24] "56aa2ad5-007d-407c-a644-48aac1e9a8f0"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $clade_name_species</span></span>
<span><span class="co">#&gt; [1] "s__Faecalibacterium_SGB15346"       "s__Faecalibacterium_prausnitzii"   </span></span>
<span><span class="co">#&gt; [3] "s__Faecalibacterium_sp_An122"       "s__Faecalibacterium_sp_CLA_AA_H233"</span></span>
<span><span class="co">#&gt; [5] "s__Faecalibacterium_sp_HTFF"       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $clade_name_terminal</span></span>
<span><span class="co">#&gt;  [1] "t__SGB15346" "t__SGB15316" "t__SGB15317" "t__SGB15318" "t__SGB15322"</span></span>
<span><span class="co">#&gt;  [6] "t__SGB15323" "t__SGB15332" "t__SGB15339" "t__SGB15342" "t__SGB15312"</span></span>
<span><span class="co">#&gt; [11] "t__SGB15315" "t__SGB15340"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="loading-into-r">Loading into R<a class="anchor" aria-label="anchor" href="#loading-into-r"></a>
</h3>
<p>We then provide our list of filtering arguments to
<code>loadParquetData</code> along with the database connection object
and the data type we are accessing and receive a Tree Summarized
Experiment object. Loading the data may take some time depending on the
file type and queries.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">basic_experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadParquetData.html">loadParquetData</a></span><span class="op">(</span>con <span class="op">=</span> <span class="va">con</span>,</span>
<span>                                    data_type <span class="op">=</span> <span class="st">"relative_abundance"</span>,</span>
<span>                                    filter_values <span class="op">=</span> <span class="va">filter_values</span>,</span>
<span>                                    custom_view <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                    include_empty_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                    dry_run <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">basic_experiment</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 9 24 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): relative_abundance</span></span>
<span><span class="co">#&gt; rownames(9):</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_SGB15346|t__SGB15346</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15316</span></span>
<span><span class="co">#&gt;   ...</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15339</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_CLA_AA_H233|t__SGB15315</span></span>
<span><span class="co">#&gt; rowData names(19): clade_name clade_name_kingdom ...</span></span>
<span><span class="co">#&gt;   NCBI_tax_id_terminal additional_species</span></span>
<span><span class="co">#&gt; colnames(24): eda61949-02dc-40ae-8dbe-bea2add85a52</span></span>
<span><span class="co">#&gt;   e47a59bb-443a-405f-9c5d-02659d80e9e5 ...</span></span>
<span><span class="co">#&gt;   09a9303d-d87d-4556-9672-04cbbcaf3d37</span></span>
<span><span class="co">#&gt;   677be4e3-722b-4e43-bd5a-36d8fbed6f86</span></span>
<span><span class="co">#&gt; colData names(524): uuid db_version ...</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_digestion_issue</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_constipation</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: NULL</span></span>
<span><span class="co">#&gt; rowTree: NULL</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
<p>Performing a dry run in order to examine the SQL query would look
much as before:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_only</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadParquetData.html">loadParquetData</a></span><span class="op">(</span>con <span class="op">=</span> <span class="va">con</span>,</span>
<span>                              data_type <span class="op">=</span> <span class="st">"relative_abundance"</span>,</span>
<span>                              filter_values <span class="op">=</span> <span class="va">filter_values</span>,</span>
<span>                              custom_view <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                              include_empty_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                              dry_run <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/explain.html" class="external-link">show_query</a></span><span class="op">(</span><span class="va">query_only</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;SQL&gt;</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">SELECT</span> q01.*</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">FROM</span> (</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">SELECT</span> relative_abundance_clade_name_species.*</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">FROM</span> relative_abundance_clade_name_species</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">WHERE</span> (clade_name_species = 's__Faecalibacterium_SGB15346')</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">UNION ALL</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">SELECT</span> relative_abundance_clade_name_species.*</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">FROM</span> relative_abundance_clade_name_species</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">WHERE</span> (clade_name_species = 's__Faecalibacterium_prausnitzii')</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">UNION ALL</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">SELECT</span> relative_abundance_clade_name_species.*</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">FROM</span> relative_abundance_clade_name_species</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">WHERE</span> (clade_name_species = 's__Faecalibacterium_sp_An122')</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">UNION ALL</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">SELECT</span> relative_abundance_clade_name_species.*</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">FROM</span> relative_abundance_clade_name_species</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">WHERE</span> (clade_name_species = 's__Faecalibacterium_sp_CLA_AA_H233')</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">UNION ALL</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">SELECT</span> relative_abundance_clade_name_species.*</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">FROM</span> relative_abundance_clade_name_species</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">WHERE</span> (clade_name_species = 's__Faecalibacterium_sp_HTFF')</span></span>
<span><span class="co">#&gt; ) q01</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">WHERE</span></span></span>
<span><span class="co">#&gt;   (clade_name_terminal IN ('t__SGB15346', 't__SGB15316', 't__SGB15317', 't__SGB15318', 't__SGB15322', 't__SGB15323', 't__SGB15332', 't__SGB15339', 't__SGB15342', 't__SGB15312', 't__SGB15315', 't__SGB15340'))<span style="color: #0000BB;"> AND</span></span></span>
<span><span class="co">#&gt;   (uuid IN ('0807eb2a-a15e-4647-8e19-2600d8fda378', 'e0fbb54f-0249-4917-a4d7-bd68acb89c62', '25172837-2849-4db3-be91-d54d6a815d00', '39ddb5e7-97f6-4d3c-812b-9653b03f99b3', '7b152a7d-e244-4e2b-b924-7195c7ecfb10', 'dd30f93b-7999-47a4-93fb-21971b899939', '1406666f-04a8-43c9-983b-4ed62fd6da4a', 'fe3de3ca-3a14-4bd8-ae1c-0dad69edc9cd', '8707e374-5ddb-4220-8cbf-364b8b0e7be1', '22848a9c-66a6-4993-9058-cb6464edb42f', '08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a', '9baef0b2-93d2-4a40-8082-d357c7f8156a', '09a9303d-d87d-4556-9672-04cbbcaf3d37', 'ac9f3532-90d8-412c-9c80-491037f0bcc2', 'eda61949-02dc-40ae-8dbe-bea2add85a52', '1f007260-be6c-4a21-800a-ad9c36129a0d', 'e47a59bb-443a-405f-9c5d-02659d80e9e5', 'b3eaf3ab-43ef-4830-ab6d-12bafed3c61e', '28f7352f-fe23-4003-93e1-41f4fedc6232', 'b07e2362-5851-4181-ba9a-15d9109ee4dd', '677be4e3-722b-4e43-bd5a-36d8fbed6f86', '0c817272-f873-475f-a401-dfe46a679a9f', '7a3945d9-21bb-434a-9a4e-bfcdeb6194de', '56aa2ad5-007d-407c-a644-48aac1e9a8f0'))</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="optional-view-customization">Optional View Customization<a class="anchor" aria-label="anchor" href="#optional-view-customization"></a>
</h3>
<p>Additionally, it is possible to use <code>dplyr</code> to preview the
data view of interest and perform more advanced functions on the data
prior to using <code>loadParquetData</code> or the
<code>collect()</code> function to load it into R. Here we use the
<code>tbl()</code> function to access the data view of choice, and pipe
it into a <code><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter()</a></code> call to select only rows that have a
value in the “additional_species” column. We can then save these calls
to a variable prior to loading them into R, and supply this to
<code>loadParquetData</code> either instead of or alongside our other
standard filtering arguments.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_filter</span> <span class="op">&lt;-</span> <span class="fu">tbl</span><span class="op">(</span><span class="va">con</span>, <span class="st">"relative_abundance_uuid"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">additional_species</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can see a preview of the data by calling the saved view:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_filter</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Source:   SQL [?? x 26]</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Database: DuckDB 1.4.2 [unknown@Linux 6.11.0-1018-azure:R 4.5.2/:memory:]</span></span></span>
<span><span class="co">#&gt;    clade_name_kingdom clade_name_phylum clade_name_class clade_name_order    </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> k__Bacteria        p__Firmicutes     c__Clostridia    o__Eubacteriales    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> k__Bacteria        p__Bacteroidota   c__Bacteroidia   o__Bacteroidales    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> k__Bacteria        p__Firmicutes     c__Bacilli       o__Lactobacillales  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> k__Bacteria        p__Bacteroidota   c__Bacteroidia   o__Bacteroidales    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> k__Bacteria        p__Bacteroidota   c__Bacteroidia   o__Bacteroidales    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> k__Bacteria        p__Firmicutes     c__Clostridia    o__Eubacteriales    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> k__Bacteria        p__Firmicutes     c__Clostridia    o__Eubacteriales    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> k__Bacteria        p__Firmicutes     c__Clostridia    o__Eubacteriales    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> k__Bacteria        p__Firmicutes     c__Clostridia    o__Eubacteriales    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> k__Bacteria        p__Actinobacteria c__Actinomycetia o__Bifidobacteriales</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 22 more variables: clade_name_family &lt;chr&gt;, clade_name_genus &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   clade_name_species &lt;chr&gt;, clade_name_terminal &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   NCBI_tax_id_kingdom &lt;chr&gt;, NCBI_tax_id_phylum &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   NCBI_tax_id_class &lt;chr&gt;, NCBI_tax_id_order &lt;chr&gt;, NCBI_tax_id_family &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   NCBI_tax_id_genus &lt;chr&gt;, NCBI_tax_id_species &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   NCBI_tax_id_terminal &lt;chr&gt;, clade_name &lt;chr&gt;, NCBI_tax_id &lt;chr&gt;, …</span></span></span></code></pre></div>
<p>Or show the query prior to running:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/explain.html" class="external-link">show_query</a></span><span class="op">(</span><span class="va">custom_filter</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;SQL&gt;</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">SELECT</span> relative_abundance_uuid.*</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">FROM</span> relative_abundance_uuid</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">WHERE</span> (NOT((additional_species IS NULL)))</span></span></code></pre></div>
<p>We then simply provide that custom view to
<code><a href="../reference/loadParquetData.html">loadParquetData()</a></code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadParquetData.html">loadParquetData</a></span><span class="op">(</span>con <span class="op">=</span> <span class="va">con</span>,</span>
<span>                                     data_type <span class="op">=</span> <span class="st">"relative_abundance"</span>,</span>
<span>                                     filter_values <span class="op">=</span> <span class="va">filter_values</span>,</span>
<span>                                     custom_view <span class="op">=</span> <span class="va">custom_filter</span>,</span>
<span>                                     include_empty_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                     dry_run <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">custom_experiment</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 7 24 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): relative_abundance</span></span>
<span><span class="co">#&gt; rownames(7):</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15316</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15332</span></span>
<span><span class="co">#&gt;   ...</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__SGB15323</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Eubacteriales|f__Oscillospiraceae|g__Faecalibacterium|s__Faecalibacterium_sp_CLA_AA_H233|t__SGB15315</span></span>
<span><span class="co">#&gt; rowData names(19): clade_name clade_name_kingdom ...</span></span>
<span><span class="co">#&gt;   NCBI_tax_id_terminal additional_species</span></span>
<span><span class="co">#&gt; colnames(24): 0807eb2a-a15e-4647-8e19-2600d8fda378</span></span>
<span><span class="co">#&gt;   08e2b754-78e2-4cb4-8ff2-95fd7b0ff44a ...</span></span>
<span><span class="co">#&gt;   e47a59bb-443a-405f-9c5d-02659d80e9e5</span></span>
<span><span class="co">#&gt;   eda61949-02dc-40ae-8dbe-bea2add85a52</span></span>
<span><span class="co">#&gt; colData names(524): uuid db_version ...</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_digestion_issue</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_constipation</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: NULL</span></span>
<span><span class="co">#&gt; rowTree: NULL</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="large-file-considerations">Large File Considerations<a class="anchor" aria-label="anchor" href="#large-file-considerations"></a>
</h3>
<p>Handling the larger data types, like “genefamilies_stratified” and
similar files, works the exact same way but must take a few factors into
consideration. There are two main limiters:</p>
<ol style="list-style-type: decimal">
<li>Hugging Face rate limits</li>
<li>Result size</li>
</ol>
<p>To avoid running into Hugging Face’s rate limiting (HTTP 429 error),
we want to make sure that all filtering arguments are very selective and
only executed on sorted columns. Using “genefamilies_stratified” as an
example, we can filter by “uuid” and “gene_family_uniref”, since those
are the two parquet files available. Filtering by “gene_family_species”
<em>while we are still accessing the parquet in a lazy manner</em> can
throw an error since a lot of the values in that column show up much
more often in the dataset. To filter these non-sorted columns, it is
recommended to use alternate filters that leverage the sorted columns
(e.g., uuid) to create the initial subset and TreeSummarizedExperiment,
then apply the less selective filters to the now locally available
data.</p>
<p>Alternatively, we can download these large files. This will
completely eliminate the rate limits since we are no longer accessing
them remotely.</p>
<p>These options do need to be balanced with your resources for storing
and handling both the raw and retrieved data, however. It may be helpful
to place a DuckDB database file in a location capable of large file
storage, or run R within a high-resource environment.</p>
<p>Below, a few example large file queries are executed. They may take a
few seconds to fully load the data.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Establish connection to genefamilies_stratified files</span></span>
<span><span class="va">con_gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/accessParquetData.html">accessParquetData</a></span><span class="op">(</span>data_types <span class="op">=</span> <span class="st">"genefamilies_stratified"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Example query: we are looking for the below gene families within a particular</span></span>
<span><span class="co"># study.</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># UniRef90_T4BVE4 - present only in SPF mice</span></span>
<span><span class="co"># UniRef90_A0A1B1SA57 - present only in WildR mice</span></span>
<span></span>
<span><span class="co"># Pull study metadata</span></span>
<span><span class="va">selected_samples</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">study_name</span> <span class="op">==</span> <span class="st">"MazmanianS_DumitrescuDG"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">select</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Do as much as possible to narrow down sample IDs prior to filtering</span></span>
<span><span class="va">wildr_ids</span> <span class="op">&lt;-</span> <span class="va">selected_samples</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">uncurated_donor_microbiome_type</span> <span class="op">==</span> <span class="st">"WildR"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">uuid</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spf_ids</span> <span class="op">&lt;-</span> <span class="va">selected_samples</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">uncurated_donor_microbiome_type</span> <span class="op">==</span> <span class="st">"SPF"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">uuid</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run the queries</span></span>
<span><span class="va">spf_ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadParquetData.html">loadParquetData</a></span><span class="op">(</span><span class="va">con_gs</span>, <span class="st">"genefamilies_stratified"</span>,</span>
<span>                          filter_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>gene_family_uniref <span class="op">=</span> <span class="st">"UniRef90_T4BVE4"</span>,</span>
<span>                                               uuid <span class="op">=</span> <span class="va">spf_ids</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">spf_ex</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 4 14 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): rpk_abundance</span></span>
<span><span class="co">#&gt; rownames(4): UniRef90_T4BVE4|g__Bacteroides.s__Bacteroides_vulgatus</span></span>
<span><span class="co">#&gt;   UniRef90_T4BVE4|unclassified</span></span>
<span><span class="co">#&gt;   UniRef90_T4BVE4|g__Bacteroides.s__Bacteroides_thetaiotaomicron</span></span>
<span><span class="co">#&gt;   UniRef90_T4BVE4|g__Parabacteroides.s__Parabacteroides_distasonis</span></span>
<span><span class="co">#&gt; rowData names(4): gene_family gene_family_uniref gene_family_genus</span></span>
<span><span class="co">#&gt;   gene_family_species</span></span>
<span><span class="co">#&gt; colnames(14): 1d949e2f-8bdb-48b7-bcf5-171c37d9ad66</span></span>
<span><span class="co">#&gt;   d2b9638a-8c15-4d1d-b5cd-3efeeeed0f2f ...</span></span>
<span><span class="co">#&gt;   01cae6dd-3b93-4abd-9f17-5fbc6a82863e</span></span>
<span><span class="co">#&gt;   ba26be3f-c2d9-4103-96e6-eacc0ff75afd</span></span>
<span><span class="co">#&gt; colData names(520): uuid humann_header ...</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_digestion_issue</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_constipation</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: NULL</span></span>
<span><span class="co">#&gt; rowTree: NULL</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span>
<span></span>
<span><span class="va">wildr_ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadParquetData.html">loadParquetData</a></span><span class="op">(</span><span class="va">con_gs</span>, <span class="st">"genefamilies_stratified"</span>,</span>
<span>                            filter_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>gene_family_uniref <span class="op">=</span> <span class="st">"UniRef90_A0A1B1SA57"</span>,</span>
<span>                                                 uuid <span class="op">=</span> <span class="va">wildr_ids</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">wildr_ex</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 2 14 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): rpk_abundance</span></span>
<span><span class="co">#&gt; rownames(2):</span></span>
<span><span class="co">#&gt;   UniRef90_A0A1B1SA57|g__Muribaculum.s__Muribaculum_intestinale</span></span>
<span><span class="co">#&gt;   UniRef90_A0A1B1SA57|unclassified</span></span>
<span><span class="co">#&gt; rowData names(4): gene_family gene_family_uniref gene_family_genus</span></span>
<span><span class="co">#&gt;   gene_family_species</span></span>
<span><span class="co">#&gt; colnames(14): 265da01f-bce1-4735-8184-43f42ba5f34b</span></span>
<span><span class="co">#&gt;   1e626f8b-ce4e-4a6a-ae29-82e7ec86b8b8 ...</span></span>
<span><span class="co">#&gt;   bfede8e1-e9fd-434f-9f74-7b1cd424edd7</span></span>
<span><span class="co">#&gt;   985f49c5-a0d2-428d-98f4-b458b0c2c0da</span></span>
<span><span class="co">#&gt; colData names(520): uuid humann_header ...</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_digestion_issue</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_constipation</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: NULL</span></span>
<span><span class="co">#&gt; rowTree: NULL</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span>
<span></span>
<span><span class="co"># The full sample ID list also works, but is a bit slower</span></span>
<span><span class="va">all_ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadParquetData.html">loadParquetData</a></span><span class="op">(</span><span class="va">con_gs</span>, <span class="st">"genefamilies_stratified"</span>,</span>
<span>                          filter_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>gene_family_uniref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UniRef90_T4BVE4"</span>, <span class="st">"UniRef90_A0A1B1SA57"</span><span class="op">)</span>,</span>
<span>                                               uuid <span class="op">=</span> <span class="va">selected_samples</span><span class="op">$</span><span class="va">uuid</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">all_ex</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 6 28 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): rpk_abundance</span></span>
<span><span class="co">#&gt; rownames(6): UniRef90_T4BVE4|g__Bacteroides.s__Bacteroides_vulgatus</span></span>
<span><span class="co">#&gt;   UniRef90_T4BVE4|unclassified ...</span></span>
<span><span class="co">#&gt;   UniRef90_A0A1B1SA57|g__Muribaculum.s__Muribaculum_intestinale</span></span>
<span><span class="co">#&gt;   UniRef90_A0A1B1SA57|unclassified</span></span>
<span><span class="co">#&gt; rowData names(4): gene_family gene_family_uniref gene_family_genus</span></span>
<span><span class="co">#&gt;   gene_family_species</span></span>
<span><span class="co">#&gt; colnames(28): 1d949e2f-8bdb-48b7-bcf5-171c37d9ad66</span></span>
<span><span class="co">#&gt;   d2b9638a-8c15-4d1d-b5cd-3efeeeed0f2f ...</span></span>
<span><span class="co">#&gt;   bfede8e1-e9fd-434f-9f74-7b1cd424edd7</span></span>
<span><span class="co">#&gt;   985f49c5-a0d2-428d-98f4-b458b0c2c0da</span></span>
<span><span class="co">#&gt; colData names(520): uuid humann_header ...</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_digestion_issue</span></span>
<span><span class="co">#&gt;   uncurated_Day_of_stool_collection_constipation</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: NULL</span></span>
<span><span class="co">#&gt; rowTree: NULL</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kaelyn Long.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
